// acorn-node/walk.js
var xtend = require('xtend')
var walk = require('acorn-walk')

var base = xtend(walk.base)
base.Import = function () ***REMOVED******REMOVED***

function simple (node, visitors, baseVisitor, state, override) ***REMOVED***
  return walk.simple(node, visitors, baseVisitor || base, state, override)
***REMOVED***

function ancestor (node, visitors, baseVisitor, state) ***REMOVED***
  return walk.ancestor(node, visitors, baseVisitor || base, state)
***REMOVED***

function recursive (node, state, funcs, baseVisitor, override) ***REMOVED***
  return walk.recursive(node, state, funcs, baseVisitor || base, override)
***REMOVED***

function full (node, callback, baseVisitor, state, override) ***REMOVED***
  return walk.full(node, callback, baseVisitor || base, state, override)
***REMOVED***

function fullAncestor (node, callback, baseVisitor, state) ***REMOVED***
  return walk.fullAncestor(node, callback, baseVisitor || base, state)
***REMOVED***

function findNodeAt (node, start, end, test, baseVisitor, state) ***REMOVED***
  return walk.findNodeAt(node, start, end, test, baseVisitor || base, state)
***REMOVED***

function findNodeAround (node, pos, test, baseVisitor, state) ***REMOVED***
  return walk.findNodeAround(node, pos, test, baseVisitor || base, state)
***REMOVED***

function findNodeAfter (node, pos, test, baseVisitor, state) ***REMOVED***
  return walk.findNodeAfter(node, pos, test, baseVisitor || base, state)
***REMOVED***

function findNodeBefore (node, pos, test, baseVisitor, state) ***REMOVED***
  return walk.findNodeBefore(node, pos, test, baseVisitor || base, state)
***REMOVED***

function make (funcs, baseVisitor) ***REMOVED***
  return walk.make(funcs, baseVisitor || base)
***REMOVED***

exports.simple = simple
exports.ancestor = ancestor
exports.recursive = recursive
exports.full = full
exports.fullAncestor = fullAncestor
exports.findNodeAt = findNodeAt
exports.findNodeAround = findNodeAround
exports.findNodeAfter = findNodeAfter
exports.findNodeBefore = findNodeBefore
exports.make = make
exports.base = base
