// browserslist/index.js
var jsReleases = require('node-releases/data/processed/envs.json')
var agents = require('caniuse-lite/dist/unpacker/agents').agents
var jsEOL = require('node-releases/data/release-schedule/release-schedule.json')
var path = require('path')
var e2c = require('electron-to-chromium/versions')

var BrowserslistError = require('./error')
var env = require('./node') // Will load browser.js in webpack

var YEAR = 365.259641 * 24 * 60 * 60 * 1000
var ANDROID_EVERGREEN_FIRST = 37

var QUERY_OR = 1
var QUERY_AND = 2

function isVersionsMatch(versionA, versionB) ***REMOVED***
  return (versionA + '.').indexOf(versionB + '.') === 0
***REMOVED***

function isEolReleased(name) ***REMOVED***
  var version = name.slice(1)
  return browserslist.nodeVersions.some(function (i) ***REMOVED***
    return isVersionsMatch(i, version)
***REMOVED***)
***REMOVED***

function normalize(versions) ***REMOVED***
  return versions.filter(function (version) ***REMOVED***
    return typeof version === 'string'
***REMOVED***)
***REMOVED***

function normalizeElectron(version) ***REMOVED***
  var versionToUse = version
  if (version.split('.').length === 3) ***REMOVED***
    versionToUse = version.split('.').slice(0, -1).join('.')
***REMOVED***
  return versionToUse
***REMOVED***

function nameMapper(name) ***REMOVED***
  return function mapName(version) ***REMOVED***
    return name + ' ' + version
***REMOVED***
***REMOVED***

function getMajor(version) ***REMOVED***
  return parseInt(version.split('.')[0])
***REMOVED***

function getMajorVersions(released, number) ***REMOVED***
  if (released.length === 0) return []
  var majorVersions = uniq(released.map(getMajor))
  var minimum = majorVersions[majorVersions.length - number]
  if (!minimum) ***REMOVED***
    return released
***REMOVED***
  var selected = []
  for (var i = released.length - 1; i >= 0; i--) ***REMOVED***
    if (minimum > getMajor(released[i])) break
    selected.unshift(released[i])
***REMOVED***
  return selected
***REMOVED***

function uniq(array) ***REMOVED***
  var filtered = []
  for (var i = 0; i < array.length; i++) ***REMOVED***
    if (filtered.indexOf(array[i]) === -1) filtered.push(array[i])
***REMOVED***
  return filtered
***REMOVED***

// Helpers

function fillUsage(result, name, data) ***REMOVED***
  for (var i in data) ***REMOVED***
    result[name + ' ' + i] = data[i]
***REMOVED***
***REMOVED***

function generateFilter(sign, version) ***REMOVED***
  version = parseFloat(version)
  if (sign === '>') ***REMOVED***
    return function (v) ***REMOVED***
      return parseFloat(v) > version
***REMOVED***
***REMOVED*** else if (sign === '>=') ***REMOVED***
    return function (v) ***REMOVED***
      return parseFloat(v) >= version
***REMOVED***
***REMOVED*** else if (sign === '<') ***REMOVED***
    return function (v) ***REMOVED***
      return parseFloat(v) < version
***REMOVED***
***REMOVED*** else ***REMOVED***
    return function (v) ***REMOVED***
      return parseFloat(v) <= version
***REMOVED***
***REMOVED***
***REMOVED***

function generateSemverFilter(sign, version) ***REMOVED***
  version = version.split('.').map(parseSimpleInt)
  version[1] = version[1] || 0
  version[2] = version[2] || 0
  if (sign === '>') ***REMOVED***
    return function (v) ***REMOVED***
      v = v.split('.').map(parseSimpleInt)
      return compareSemver(v, version) > 0
***REMOVED***
***REMOVED*** else if (sign === '>=') ***REMOVED***
    return function (v) ***REMOVED***
      v = v.split('.').map(parseSimpleInt)
      return compareSemver(v, version) >= 0
***REMOVED***
***REMOVED*** else if (sign === '<') ***REMOVED***
    return function (v) ***REMOVED***
      v = v.split('.').map(parseSimpleInt)
      return compareSemver(version, v) > 0
***REMOVED***
***REMOVED*** else ***REMOVED***
    return function (v) ***REMOVED***
      v = v.split('.').map(parseSimpleInt)
      return compareSemver(version, v) >= 0
***REMOVED***
***REMOVED***
***REMOVED***

function parseSimpleInt(x) ***REMOVED***
  return parseInt(x)
***REMOVED***

function compare(a, b) ***REMOVED***
  if (a < b) return -1
  if (a > b) return +1
  return 0
***REMOVED***

function compareSemver(a, b) ***REMOVED***
  return (
    compare(parseInt(a[0]), parseInt(b[0])) ||
    compare(parseInt(a[1] || '0'), parseInt(b[1] || '0')) ||
    compare(parseInt(a[2] || '0'), parseInt(b[2] || '0'))
  )
***REMOVED***

// this follows the npm-like semver behavior
function semverFilterLoose(operator, range) ***REMOVED***
  range = range.split('.').map(parseSimpleInt)
  if (typeof range[1] === 'undefined') ***REMOVED***
    range[1] = 'x'
***REMOVED***
  // ignore any patch version because we only return minor versions
  // range[2] = 'x'
  switch (operator) ***REMOVED***
    case '<=':
      return function (version) ***REMOVED***
        version = version.split('.').map(parseSimpleInt)
        return compareSemverLoose(version, range) <= 0
  ***REMOVED***
    case '>=':
    default:
      return function (version) ***REMOVED***
        version = version.split('.').map(parseSimpleInt)
        return compareSemverLoose(version, range) >= 0
  ***REMOVED***
***REMOVED***
***REMOVED***

// this follows the npm-like semver behavior
function compareSemverLoose(version, range) ***REMOVED***
  if (version[0] !== range[0]) ***REMOVED***
    return version[0] < range[0] ? -1 : +1
***REMOVED***
  if (range[1] === 'x') ***REMOVED***
    return 0
***REMOVED***
  if (version[1] !== range[1]) ***REMOVED***
    return version[1] < range[1] ? -1 : +1
***REMOVED***
  return 0
***REMOVED***

function resolveVersion(data, version) ***REMOVED***
  if (data.versions.indexOf(version) !== -1) ***REMOVED***
    return version
***REMOVED*** else if (browserslist.versionAliases[data.name][version]) ***REMOVED***
    return browserslist.versionAliases[data.name][version]
***REMOVED*** else ***REMOVED***
    return false
***REMOVED***
***REMOVED***

function normalizeVersion(data, version) ***REMOVED***
  var resolved = resolveVersion(data, version)
  if (resolved) ***REMOVED***
    return resolved
***REMOVED*** else if (data.versions.length === 1) ***REMOVED***
    return data.versions[0]
***REMOVED*** else ***REMOVED***
    return false
***REMOVED***
***REMOVED***

function filterByYear(since, context) ***REMOVED***
  since = since / 1000
  return Object.keys(agents).reduce(function (selected, name) ***REMOVED***
    var data = byName(name, context)
    if (!data) return selected
    var versions = Object.keys(data.releaseDate).filter(function (v) ***REMOVED***
      var date = data.releaseDate[v]
      return date !== null && date >= since
***REMOVED***)
    return selected.concat(versions.map(nameMapper(data.name)))
***REMOVED*** [])
***REMOVED***

function cloneData(data) ***REMOVED***
  return ***REMOVED***
    name: data.name,
    versions: data.versions,
    released: data.released,
    releaseDate: data.releaseDate
***REMOVED***
***REMOVED***

function mapVersions(data, map) ***REMOVED***
  data.versions = data.versions.map(function (i) ***REMOVED***
    return map[i] || i
***REMOVED***)
  data.released = data.versions.map(function (i) ***REMOVED***
    return map[i] || i
***REMOVED***)
  var fixedDate = ***REMOVED******REMOVED***
  for (var i in data.releaseDate) ***REMOVED***
    fixedDate[map[i] || i] = data.releaseDate[i]
***REMOVED***
  data.releaseDate = fixedDate
  return data
***REMOVED***

function byName(name, context) ***REMOVED***
  name = name.toLowerCase()
  name = browserslist.aliases[name] || name
  if (context.mobileToDesktop && browserslist.desktopNames[name]) ***REMOVED***
    var desktop = browserslist.data[browserslist.desktopNames[name]]
    if (name === 'android') ***REMOVED***
      return normalizeAndroidData(cloneData(browserslist.data[name]), desktop)
***REMOVED*** else ***REMOVED***
      var cloned = cloneData(desktop)
      cloned.name = name
      if (name === 'op_mob') ***REMOVED***
        cloned = mapVersions(cloned, ***REMOVED*** '10.0-10.1': '10' ***REMOVED***)
  ***REMOVED***
      return cloned
***REMOVED***
***REMOVED***
  return browserslist.data[name]
***REMOVED***

function normalizeAndroidVersions(androidVersions, chromeVersions) ***REMOVED***
  var firstEvergreen = ANDROID_EVERGREEN_FIRST
  var last = chromeVersions[chromeVersions.length - 1]
  return androidVersions
    .filter(function (version) ***REMOVED***
      return /^(?:[2-4]\.|[34]$)/.test(version)
***REMOVED***)
    .concat(chromeVersions.slice(firstEvergreen - last - 1))
***REMOVED***

function normalizeAndroidData(android, chrome) ***REMOVED***
  android.released = normalizeAndroidVersions(android.released, chrome.released)
  android.versions = normalizeAndroidVersions(android.versions, chrome.versions)
  return android
***REMOVED***

function checkName(name, context) ***REMOVED***
  var data = byName(name, context)
  if (!data) throw new BrowserslistError('Unknown browser ' + name)
  return data
***REMOVED***

function unknownQuery(query) ***REMOVED***
  return new BrowserslistError(
    'Unknown browser query `' +
      query +
      '`. ' +
      'Maybe you are using old Browserslist or made typo in query.'
  )
***REMOVED***

function filterAndroid(list, versions, context) ***REMOVED***
  if (context.mobileToDesktop) return list
  var released = browserslist.data.android.released
  var last = released[released.length - 1]
  var diff = last - ANDROID_EVERGREEN_FIRST - versions
  if (diff > 0) ***REMOVED***
    return list.slice(-1)
***REMOVED*** else ***REMOVED***
    return list.slice(diff - 1)
***REMOVED***
***REMOVED***

/**
 * Resolves queries into a browser list.
 * @param ***REMOVED***string|string[]***REMOVED*** queries Queries to combine.
 * Either an array of queries or a long string of queries.
 * @param ***REMOVED***object***REMOVED*** [context] Optional arguments to
 * the select function in `queries`.
 * @returns ***REMOVED***string[]***REMOVED*** A list of browsers
 */
function resolve(queries, context) ***REMOVED***
  if (Array.isArray(queries)) ***REMOVED***
    queries = flatten(queries.map(parse))
***REMOVED*** else ***REMOVED***
    queries = parse(queries)
***REMOVED***

  return queries.reduce(function (result, query, index) ***REMOVED***
    var selection = query.queryString

    var isExclude = selection.indexOf('not ') === 0
    if (isExclude) ***REMOVED***
      if (index === 0) ***REMOVED***
        throw new BrowserslistError(
          'Write any browsers query (for instance, `defaults`) ' +
            'before `' +
            selection +
            '`'
        )
  ***REMOVED***
      selection = selection.slice(4)
***REMOVED***

    for (var i = 0; i < QUERIES.length; i++) ***REMOVED***
      var type = QUERIES[i]
      var match = selection.match(type.regexp)
      if (match) ***REMOVED***
        var args = [context].concat(match.slice(1))
        var array = type.select.apply(browserslist, args).map(function (j) ***REMOVED***
          var parts = j.split(' ')
          if (parts[1] === '0') ***REMOVED***
            return parts[0] + ' ' + byName(parts[0], context).versions[0]
      ***REMOVED*** else ***REMOVED***
            return j
      ***REMOVED***
    ***REMOVED***)

        switch (query.type) ***REMOVED***
          case QUERY_AND:
            if (isExclude) ***REMOVED***
              return result.filter(function (j) ***REMOVED***
                return array.indexOf(j) === -1
          ***REMOVED***)
        ***REMOVED*** else ***REMOVED***
              return result.filter(function (j) ***REMOVED***
                return array.indexOf(j) !== -1
          ***REMOVED***)
        ***REMOVED***
          case QUERY_OR:
          default:
            if (isExclude) ***REMOVED***
              var filter = ***REMOVED******REMOVED***
              array.forEach(function (j) ***REMOVED***
                filter[j] = true
          ***REMOVED***)
              return result.filter(function (j) ***REMOVED***
                return !filter[j]
          ***REMOVED***)
        ***REMOVED***
            return result.concat(array)
    ***REMOVED***
  ***REMOVED***
***REMOVED***

    throw unknownQuery(selection)
***REMOVED*** [])
***REMOVED***

var cache = ***REMOVED******REMOVED***

/**
 * Return array of browsers by selection queries.
 *
 * @param ***REMOVED***(string|string[])***REMOVED*** [queries=browserslist.defaults] Browser queries.
 * @param ***REMOVED***object***REMOVED*** [opts] Options.
 * @param ***REMOVED***string***REMOVED*** [opts.path="."] Path to processed file.
 *                                 It will be used to find config files.
 * @param ***REMOVED***string***REMOVED*** [opts.env="production"] Processing environment.
 *                                         It will be used to take right
 *                                         queries from config file.
 * @param ***REMOVED***string***REMOVED*** [opts.config] Path to config file with queries.
 * @param ***REMOVED***object***REMOVED*** [opts.stats] Custom browser usage statistics
 *                              for "> 1% in my stats" query.
 * @param ***REMOVED***boolean***REMOVED*** [opts.ignoreUnknownVersions=false] Do not throw on unknown
 *                                                     version in direct query.
 * @param ***REMOVED***boolean***REMOVED*** [opts.dangerousExtend] Disable security checks
 *                                         for extend query.
 * @param ***REMOVED***boolean***REMOVED*** [opts.throwOnMissing] Throw error on missing env.
 * @param ***REMOVED***boolean***REMOVED*** [opts.mobileToDesktop] Alias mobile browsers to the desktop
 *                                         version when Can I Use doesn't have
 *                                         data about the specified version.
 * @returns ***REMOVED***string[]***REMOVED*** Array with browser names in Can I Use.
 *
 * @example
 * browserslist('IE >= 10, IE 8') //=> ['ie 11', 'ie 10', 'ie 8']
 */
function browserslist(queries, opts) ***REMOVED***
  if (typeof opts === 'undefined') opts = ***REMOVED******REMOVED***

  if (typeof opts.path === 'undefined') ***REMOVED***
    opts.path = path.resolve ? path.resolve('.') : '.'
***REMOVED***

  if (typeof queries === 'undefined' || queries === null) ***REMOVED***
    var config = browserslist.loadConfig(opts)
    if (config) ***REMOVED***
      queries = config
***REMOVED*** else ***REMOVED***
      queries = browserslist.defaults
***REMOVED***
***REMOVED***

  if (!(typeof queries === 'string' || Array.isArray(queries))) ***REMOVED***
    throw new BrowserslistError(
      'Browser queries must be an array or string. Got ' + typeof queries + '.'
    )
***REMOVED***

  var context = ***REMOVED***
    ignoreUnknownVersions: opts.ignoreUnknownVersions,
    dangerousExtend: opts.dangerousExtend,
    mobileToDesktop: opts.mobileToDesktop,
    path: opts.path,
    env: opts.env
***REMOVED***

  env.oldDataWarning(browserslist.data)
  var stats = env.getStat(opts, browserslist.data)
  if (stats) ***REMOVED***
    context.customUsage = ***REMOVED******REMOVED***
    for (var browser in stats) ***REMOVED***
      fillUsage(context.customUsage, browser, stats[browser])
***REMOVED***
***REMOVED***

  var cacheKey = JSON.stringify([queries, context])
  if (cache[cacheKey]) return cache[cacheKey]

  var result = uniq(resolve(queries, context)).sort(function (name1, name2) ***REMOVED***
    name1 = name1.split(' ')
    name2 = name2.split(' ')
    if (name1[0] === name2[0]) ***REMOVED***
      // assumptions on caniuse data
      // 1) version ranges never overlaps
      // 2) if version is not a range, it never contains `-`
      var version1 = name1[1].split('-')[0]
      var version2 = name2[1].split('-')[0]
      return compareSemver(version2.split('.'), version1.split('.'))
***REMOVED*** else ***REMOVED***
      return compare(name1[0], name2[0])
***REMOVED***
***REMOVED***)
  if (!process.env.BROWSERSLIST_DISABLE_CACHE) ***REMOVED***
    cache[cacheKey] = result
***REMOVED***
  return result
***REMOVED***

function parse(queries) ***REMOVED***
  var qs = []
  do ***REMOVED***
    queries = doMatch(queries, qs)
***REMOVED*** while (queries)
  return qs
***REMOVED***

function doMatch(string, qs) ***REMOVED***
  var or = /^(?:,\s*|\s+or\s+)(.*)/i
  var and = /^\s+and\s+(.*)/i

  return find(string, function (parsed, n, max) ***REMOVED***
    if (and.test(parsed)) ***REMOVED***
      qs.unshift(***REMOVED*** type: QUERY_AND, queryString: parsed.match(and)[1] ***REMOVED***)
      return true
***REMOVED*** else if (or.test(parsed)) ***REMOVED***
      qs.unshift(***REMOVED*** type: QUERY_OR, queryString: parsed.match(or)[1] ***REMOVED***)
      return true
***REMOVED*** else if (n === max) ***REMOVED***
      qs.unshift(***REMOVED*** type: QUERY_OR, queryString: parsed.trim() ***REMOVED***)
      return true
***REMOVED***
    return false
***REMOVED***)
***REMOVED***

function find(string, predicate) ***REMOVED***
  for (var n = 1, max = string.length; n <= max; n++) ***REMOVED***
    var parsed = string.substr(-n, n)
    if (predicate(parsed, n, max)) ***REMOVED***
      return string.slice(0, -n)
***REMOVED***
***REMOVED***
  return ''
***REMOVED***

function flatten(array) ***REMOVED***
  if (!Array.isArray(array)) return [array]
  return array.reduce(function (a, b) ***REMOVED***
    return a.concat(flatten(b))
***REMOVED*** [])
***REMOVED***

// Will be filled by Can I Use data below
browserslist.cache = ***REMOVED******REMOVED***
browserslist.data = ***REMOVED******REMOVED***
browserslist.usage = ***REMOVED***
  global: ***REMOVED******REMOVED***,
  custom: null
***REMOVED***

// Default browsers query
browserslist.defaults = ['> 0.5%', 'last 2 versions', 'Firefox ESR', 'not dead']

// Browser names aliases
browserslist.aliases = ***REMOVED***
  fx: 'firefox',
  ff: 'firefox',
  ios: 'ios_saf',
  explorer: 'ie',
  blackberry: 'bb',
  explorermobile: 'ie_mob',
  operamini: 'op_mini',
  operamobile: 'op_mob',
  chromeandroid: 'and_chr',
  firefoxandroid: 'and_ff',
  ucandroid: 'and_uc',
  qqandroid: 'and_qq'
***REMOVED***

// Can I Use only provides a few versions for some browsers (e.g. and_chr).
// Fallback to a similar browser for unknown versions
browserslist.desktopNames = ***REMOVED***
  and_chr: 'chrome',
  and_ff: 'firefox',
  ie_mob: 'ie',
  op_mob: 'opera',
  android: 'chrome' // has extra processing logic
***REMOVED***

// Aliases to work with joined versions like `ios_saf 7.0-7.1`
browserslist.versionAliases = ***REMOVED******REMOVED***

browserslist.clearCaches = env.clearCaches
browserslist.parseConfig = env.parseConfig
browserslist.readConfig = env.readConfig
browserslist.findConfig = env.findConfig
browserslist.loadConfig = env.loadConfig

/**
 * Return browsers market coverage.
 *
 * @param ***REMOVED***string[]***REMOVED*** browsers Browsers names in Can I Use.
 * @param ***REMOVED***string|object***REMOVED*** [stats="global"] Which statistics should be used.
 *                                         Country code or custom statistics.
 *                                         Pass `"my stats"` to load statistics
 *                                         from Browserslist files.
 *
 * @return ***REMOVED***number***REMOVED*** Total market coverage for all selected browsers.
 *
 * @example
 * browserslist.coverage(browserslist('> 1% in US'), 'US') //=> 83.1
 */
browserslist.coverage = function (browsers, stats) ***REMOVED***
  var data
  if (typeof stats === 'undefined') ***REMOVED***
    data = browserslist.usage.global
***REMOVED*** else if (stats === 'my stats') ***REMOVED***
    var opts = ***REMOVED******REMOVED***
    opts.path = path.resolve ? path.resolve('.') : '.'
    var customStats = env.getStat(opts)
    if (!customStats) ***REMOVED***
      throw new BrowserslistError('Custom usage statistics was not provided')
***REMOVED***
    data = ***REMOVED******REMOVED***
    for (var browser in customStats) ***REMOVED***
      fillUsage(data, browser, customStats[browser])
***REMOVED***
***REMOVED*** else if (typeof stats === 'string') ***REMOVED***
    if (stats.length > 2) ***REMOVED***
      stats = stats.toLowerCase()
***REMOVED*** else ***REMOVED***
      stats = stats.toUpperCase()
***REMOVED***
    env.loadCountry(browserslist.usage, stats, browserslist.data)
    data = browserslist.usage[stats]
***REMOVED*** else ***REMOVED***
    if ('dataByBrowser' in stats) ***REMOVED***
      stats = stats.dataByBrowser
***REMOVED***
    data = ***REMOVED******REMOVED***
    for (var name in stats) ***REMOVED***
      for (var version in stats[name]) ***REMOVED***
        data[name + ' ' + version] = stats[name][version]
  ***REMOVED***
***REMOVED***
***REMOVED***

  return browsers.reduce(function (all, i) ***REMOVED***
    var usage = data[i]
    if (usage === undefined) ***REMOVED***
      usage = data[i.replace(/ \S+$/, ' 0')]
***REMOVED***
    return all + (usage || 0)
***REMOVED*** 0)
***REMOVED***

function nodeQuery(context, version) ***REMOVED***
  var matched = browserslist.nodeVersions.filter(function (i) ***REMOVED***
    return isVersionsMatch(i, version)
***REMOVED***)
  if (matched.length === 0) ***REMOVED***
    if (context.ignoreUnknownVersions) ***REMOVED***
      return []
***REMOVED*** else ***REMOVED***
      throw new BrowserslistError('Unknown version ' + version + ' of Node.js')
***REMOVED***
***REMOVED***
  return ['node ' + matched[matched.length - 1]]
***REMOVED***

function sinceQuery(context, year, month, date) ***REMOVED***
  year = parseInt(year)
  month = parseInt(month || '01') - 1
  date = parseInt(date || '01')
  return filterByYear(Date.UTC(year, month, date, 0, 0, 0), context)
***REMOVED***

function coverQuery(context, coverage, statMode) ***REMOVED***
  coverage = parseFloat(coverage)
  var usage = browserslist.usage.global
  if (statMode) ***REMOVED***
    if (statMode.match(/^my\s+stats$/i)) ***REMOVED***
      if (!context.customUsage) ***REMOVED***
        throw new BrowserslistError('Custom usage statistics was not provided')
  ***REMOVED***
      usage = context.customUsage
***REMOVED*** else ***REMOVED***
      var place
      if (statMode.length === 2) ***REMOVED***
        place = statMode.toUpperCase()
  ***REMOVED*** else ***REMOVED***
        place = statMode.toLowerCase()
  ***REMOVED***
      env.loadCountry(browserslist.usage, place, browserslist.data)
      usage = browserslist.usage[place]
***REMOVED***
***REMOVED***
  var versions = Object.keys(usage).sort(function (a, b) ***REMOVED***
    return usage[b] - usage[a]
***REMOVED***)
  var coveraged = 0
  var result = []
  var version
  for (var i = 0; i < versions.length; i++) ***REMOVED***
    version = versions[i]
    if (usage[version] === 0) break
    coveraged += usage[version]
    result.push(version)
    if (coveraged >= coverage) break
***REMOVED***
  return result
***REMOVED***

var QUERIES = [
  ***REMOVED***
    regexp: /^last\s+(\d+)\s+major\s+versions?$/i,
    select: function (context, versions) ***REMOVED***
      return Object.keys(agents).reduce(function (selected, name) ***REMOVED***
        var data = byName(name, context)
        if (!data) return selected
        var list = getMajorVersions(data.released, versions)
        list = list.map(nameMapper(data.name))
        if (data.name === 'android') ***REMOVED***
          list = filterAndroid(list, versions, context)
    ***REMOVED***
        return selected.concat(list)
  ***REMOVED***, [])
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^last\s+(\d+)\s+versions?$/i,
    select: function (context, versions) ***REMOVED***
      return Object.keys(agents).reduce(function (selected, name) ***REMOVED***
        var data = byName(name, context)
        if (!data) return selected
        var list = data.released.slice(-versions)
        list = list.map(nameMapper(data.name))
        if (data.name === 'android') ***REMOVED***
          list = filterAndroid(list, versions, context)
    ***REMOVED***
        return selected.concat(list)
  ***REMOVED***, [])
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^last\s+(\d+)\s+electron\s+major\s+versions?$/i,
    select: function (context, versions) ***REMOVED***
      var validVersions = getMajorVersions(Object.keys(e2c), versions)
      return validVersions.map(function (i) ***REMOVED***
        return 'chrome ' + e2c[i]
  ***REMOVED***)
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^last\s+(\d+)\s+node\s+major\s+versions?$/i,
    select: function (context, versions) ***REMOVED***
      return getMajorVersions(browserslist.nodeVersions, versions).map(
        function (version) ***REMOVED***
          return 'node ' + version
    ***REMOVED***
      )
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^last\s+(\d+)\s+(\w+)\s+major\s+versions?$/i,
    select: function (context, versions, name) ***REMOVED***
      var data = checkName(name, context)
      var validVersions = getMajorVersions(data.released, versions)
      var list = validVersions.map(nameMapper(data.name))
      if (data.name === 'android') ***REMOVED***
        list = filterAndroid(list, versions, context)
  ***REMOVED***
      return list
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^last\s+(\d+)\s+electron\s+versions?$/i,
    select: function (context, versions) ***REMOVED***
      return Object.keys(e2c)
        .slice(-versions)
        .map(function (i) ***REMOVED***
          return 'chrome ' + e2c[i]
    ***REMOVED***)
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^last\s+(\d+)\s+node\s+versions?$/i,
    select: function (context, versions) ***REMOVED***
      return browserslist.nodeVersions.slice(-versions).map(function (version) ***REMOVED***
        return 'node ' + version
  ***REMOVED***)
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^last\s+(\d+)\s+(\w+)\s+versions?$/i,
    select: function (context, versions, name) ***REMOVED***
      var data = checkName(name, context)
      var list = data.released.slice(-versions).map(nameMapper(data.name))
      if (data.name === 'android') ***REMOVED***
        list = filterAndroid(list, versions, context)
  ***REMOVED***
      return list
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^unreleased\s+versions$/i,
    select: function (context) ***REMOVED***
      return Object.keys(agents).reduce(function (selected, name) ***REMOVED***
        var data = byName(name, context)
        if (!data) return selected
        var list = data.versions.filter(function (v) ***REMOVED***
          return data.released.indexOf(v) === -1
    ***REMOVED***)
        list = list.map(nameMapper(data.name))
        return selected.concat(list)
  ***REMOVED***, [])
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^unreleased\s+electron\s+versions?$/i,
    select: function () ***REMOVED***
      return []
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^unreleased\s+(\w+)\s+versions?$/i,
    select: function (context, name) ***REMOVED***
      var data = checkName(name, context)
      return data.versions
        .filter(function (v) ***REMOVED***
          return data.released.indexOf(v) === -1
    ***REMOVED***)
        .map(nameMapper(data.name))
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^last\s+(\d*.?\d+)\s+years?$/i,
    select: function (context, years) ***REMOVED***
      return filterByYear(Date.now() - YEAR * years, context)
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^since (\d+)$/i,
    select: sinceQuery
***REMOVED***
  ***REMOVED***
    regexp: /^since (\d+)-(\d+)$/i,
    select: sinceQuery
***REMOVED***
  ***REMOVED***
    regexp: /^since (\d+)-(\d+)-(\d+)$/i,
    select: sinceQuery
***REMOVED***
  ***REMOVED***
    regexp: /^(>=?|<=?)\s*(\d+|\d+\.\d+|\.\d+)%$/,
    select: function (context, sign, popularity) ***REMOVED***
      popularity = parseFloat(popularity)
      var usage = browserslist.usage.global
      return Object.keys(usage).reduce(function (result, version) ***REMOVED***
        if (sign === '>') ***REMOVED***
          if (usage[version] > popularity) ***REMOVED***
            result.push(version)
      ***REMOVED***
    ***REMOVED*** else if (sign === '<') ***REMOVED***
          if (usage[version] < popularity) ***REMOVED***
            result.push(version)
      ***REMOVED***
    ***REMOVED*** else if (sign === '<=') ***REMOVED***
          if (usage[version] <= popularity) ***REMOVED***
            result.push(version)
      ***REMOVED***
    ***REMOVED*** else if (usage[version] >= popularity) ***REMOVED***
          result.push(version)
    ***REMOVED***
        return result
  ***REMOVED***, [])
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^(>=?|<=?)\s*(\d+|\d+\.\d+|\.\d+)%\s+in\s+my\s+stats$/,
    select: function (context, sign, popularity) ***REMOVED***
      popularity = parseFloat(popularity)
      if (!context.customUsage) ***REMOVED***
        throw new BrowserslistError('Custom usage statistics was not provided')
  ***REMOVED***
      var usage = context.customUsage
      return Object.keys(usage).reduce(function (result, version) ***REMOVED***
        var percentage = usage[version]
        if (percentage == null) ***REMOVED***
          return result
    ***REMOVED***

        if (sign === '>') ***REMOVED***
          if (percentage > popularity) ***REMOVED***
            result.push(version)
      ***REMOVED***
    ***REMOVED*** else if (sign === '<') ***REMOVED***
          if (percentage < popularity) ***REMOVED***
            result.push(version)
      ***REMOVED***
    ***REMOVED*** else if (sign === '<=') ***REMOVED***
          if (percentage <= popularity) ***REMOVED***
            result.push(version)
      ***REMOVED***
    ***REMOVED*** else if (percentage >= popularity) ***REMOVED***
          result.push(version)
    ***REMOVED***
        return result
  ***REMOVED***, [])
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^(>=?|<=?)\s*(\d+|\d+\.\d+|\.\d+)%\s+in\s+(\S+)\s+stats$/,
    select: function (context, sign, popularity, name) ***REMOVED***
      popularity = parseFloat(popularity)
      var stats = env.loadStat(context, name, browserslist.data)
      if (stats) ***REMOVED***
        context.customUsage = ***REMOVED******REMOVED***
        for (var browser in stats) ***REMOVED***
          fillUsage(context.customUsage, browser, stats[browser])
    ***REMOVED***
  ***REMOVED***
      if (!context.customUsage) ***REMOVED***
        throw new BrowserslistError('Custom usage statistics was not provided')
  ***REMOVED***
      var usage = context.customUsage
      return Object.keys(usage).reduce(function (result, version) ***REMOVED***
        var percentage = usage[version]
        if (percentage == null) ***REMOVED***
          return result
    ***REMOVED***

        if (sign === '>') ***REMOVED***
          if (percentage > popularity) ***REMOVED***
            result.push(version)
      ***REMOVED***
    ***REMOVED*** else if (sign === '<') ***REMOVED***
          if (percentage < popularity) ***REMOVED***
            result.push(version)
      ***REMOVED***
    ***REMOVED*** else if (sign === '<=') ***REMOVED***
          if (percentage <= popularity) ***REMOVED***
            result.push(version)
      ***REMOVED***
    ***REMOVED*** else if (percentage >= popularity) ***REMOVED***
          result.push(version)
    ***REMOVED***
        return result
  ***REMOVED***, [])
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^(>=?|<=?)\s*(\d+|\d+\.\d+|\.\d+)%\s+in\s+((alt-)?\w\w)$/,
    select: function (context, sign, popularity, place) ***REMOVED***
      popularity = parseFloat(popularity)
      if (place.length === 2) ***REMOVED***
        place = place.toUpperCase()
  ***REMOVED*** else ***REMOVED***
        place = place.toLowerCase()
  ***REMOVED***
      env.loadCountry(browserslist.usage, place, browserslist.data)
      var usage = browserslist.usage[place]
      return Object.keys(usage).reduce(function (result, version) ***REMOVED***
        var percentage = usage[version]
        if (percentage == null) ***REMOVED***
          return result
    ***REMOVED***

        if (sign === '>') ***REMOVED***
          if (percentage > popularity) ***REMOVED***
            result.push(version)
      ***REMOVED***
    ***REMOVED*** else if (sign === '<') ***REMOVED***
          if (percentage < popularity) ***REMOVED***
            result.push(version)
      ***REMOVED***
    ***REMOVED*** else if (sign === '<=') ***REMOVED***
          if (percentage <= popularity) ***REMOVED***
            result.push(version)
      ***REMOVED***
    ***REMOVED*** else if (percentage >= popularity) ***REMOVED***
          result.push(version)
    ***REMOVED***
        return result
  ***REMOVED***, [])
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^cover\s+(\d+|\d+\.\d+|\.\d+)%$/i,
    select: coverQuery
***REMOVED***
  ***REMOVED***
    regexp: /^cover\s+(\d+|\d+\.\d+|\.\d+)%\s+in\s+(my\s+stats|(alt-)?\w\w)$/i,
    select: coverQuery
***REMOVED***
  ***REMOVED***
    regexp: /^supports\s+([\w-]+)$/,
    select: function (context, feature) ***REMOVED***
      env.loadFeature(browserslist.cache, feature)
      var features = browserslist.cache[feature]
      return Object.keys(features).reduce(function (result, version) ***REMOVED***
        var flags = features[version]
        if (flags.indexOf('y') >= 0 || flags.indexOf('a') >= 0) ***REMOVED***
          result.push(version)
    ***REMOVED***
        return result
  ***REMOVED***, [])
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^electron\s+([\d.]+)\s*-\s*([\d.]+)$/i,
    select: function (context, from, to) ***REMOVED***
      var fromToUse = normalizeElectron(from)
      var toToUse = normalizeElectron(to)
      if (!e2c[fromToUse]) ***REMOVED***
        throw new BrowserslistError('Unknown version ' + from + ' of electron')
  ***REMOVED***
      if (!e2c[toToUse]) ***REMOVED***
        throw new BrowserslistError('Unknown version ' + to + ' of electron')
  ***REMOVED***
      from = parseFloat(from)
      to = parseFloat(to)
      return Object.keys(e2c)
        .filter(function (i) ***REMOVED***
          var parsed = parseFloat(i)
          return parsed >= from && parsed <= to
    ***REMOVED***)
        .map(function (i) ***REMOVED***
          return 'chrome ' + e2c[i]
    ***REMOVED***)
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^node\s+([\d.]+)\s*-\s*([\d.]+)$/i,
    select: function (context, from, to) ***REMOVED***
      return browserslist.nodeVersions
        .filter(semverFilterLoose('>=', from))
        .filter(semverFilterLoose('<=', to))
        .map(function (v) ***REMOVED***
          return 'node ' + v
    ***REMOVED***)
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^(\w+)\s+([\d.]+)\s*-\s*([\d.]+)$/i,
    select: function (context, name, from, to) ***REMOVED***
      var data = checkName(name, context)
      from = parseFloat(normalizeVersion(data, from) || from)
      to = parseFloat(normalizeVersion(data, to) || to)
      function filter(v) ***REMOVED***
        var parsed = parseFloat(v)
        return parsed >= from && parsed <= to
  ***REMOVED***
      return data.released.filter(filter).map(nameMapper(data.name))
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^electron\s*(>=?|<=?)\s*([\d.]+)$/i,
    select: function (context, sign, version) ***REMOVED***
      var versionToUse = normalizeElectron(version)
      return Object.keys(e2c)
        .filter(generateFilter(sign, versionToUse))
        .map(function (i) ***REMOVED***
          return 'chrome ' + e2c[i]
    ***REMOVED***)
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^node\s*(>=?|<=?)\s*([\d.]+)$/i,
    select: function (context, sign, version) ***REMOVED***
      return browserslist.nodeVersions
        .filter(generateSemverFilter(sign, version))
        .map(function (v) ***REMOVED***
          return 'node ' + v
    ***REMOVED***)
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^(\w+)\s*(>=?|<=?)\s*([\d.]+)$/,
    select: function (context, name, sign, version) ***REMOVED***
      var data = checkName(name, context)
      var alias = browserslist.versionAliases[data.name][version]
      if (alias) ***REMOVED***
        version = alias
  ***REMOVED***
      return data.released
        .filter(generateFilter(sign, version))
        .map(function (v) ***REMOVED***
          return data.name + ' ' + v
    ***REMOVED***)
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^(firefox|ff|fx)\s+esr$/i,
    select: function () ***REMOVED***
      return ['firefox 91']
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /(operamini|op_mini)\s+all/i,
    select: function () ***REMOVED***
      return ['op_mini all']
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^electron\s+([\d.]+)$/i,
    select: function (context, version) ***REMOVED***
      var versionToUse = normalizeElectron(version)
      var chrome = e2c[versionToUse]
      if (!chrome) ***REMOVED***
        throw new BrowserslistError(
          'Unknown version ' + version + ' of electron'
        )
  ***REMOVED***
      return ['chrome ' + chrome]
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^node\s+(\d+)$/i,
    select: nodeQuery
***REMOVED***
  ***REMOVED***
    regexp: /^node\s+(\d+\.\d+)$/i,
    select: nodeQuery
***REMOVED***
  ***REMOVED***
    regexp: /^node\s+(\d+\.\d+\.\d+)$/i,
    select: nodeQuery
***REMOVED***
  ***REMOVED***
    regexp: /^current\s+node$/i,
    select: function (context) ***REMOVED***
      return [env.currentNode(resolve, context)]
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^maintained\s+node\s+versions$/i,
    select: function (context) ***REMOVED***
      var now = Date.now()
      var queries = Object.keys(jsEOL)
        .filter(function (key) ***REMOVED***
          return (
            now < Date.parse(jsEOL[key].end) &&
            now > Date.parse(jsEOL[key].start) &&
            isEolReleased(key)
          )
    ***REMOVED***)
        .map(function (key) ***REMOVED***
          return 'node ' + key.slice(1)
    ***REMOVED***)
      return resolve(queries, context)
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^phantomjs\s+1.9$/i,
    select: function () ***REMOVED***
      return ['safari 5']
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^phantomjs\s+2.1$/i,
    select: function () ***REMOVED***
      return ['safari 6']
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^(\w+)\s+(tp|[\d.]+)$/i,
    select: function (context, name, version) ***REMOVED***
      if (/^tp$/i.test(version)) version = 'TP'
      var data = checkName(name, context)
      var alias = normalizeVersion(data, version)
      if (alias) ***REMOVED***
        version = alias
  ***REMOVED*** else ***REMOVED***
        if (version.indexOf('.') === -1) ***REMOVED***
          alias = version + '.0'
    ***REMOVED*** else ***REMOVED***
          alias = version.replace(/\.0$/, '')
    ***REMOVED***
        alias = normalizeVersion(data, alias)
        if (alias) ***REMOVED***
          version = alias
    ***REMOVED*** else if (context.ignoreUnknownVersions) ***REMOVED***
          return []
    ***REMOVED*** else ***REMOVED***
          throw new BrowserslistError(
            'Unknown version ' + version + ' of ' + name
          )
    ***REMOVED***
  ***REMOVED***
      return [data.name + ' ' + version]
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^browserslist config$/i,
    select: function (context) ***REMOVED***
      return browserslist(undefined, context)
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^extends (.+)$/i,
    select: function (context, name) ***REMOVED***
      return resolve(env.loadQueries(context, name), context)
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^defaults$/i,
    select: function (context) ***REMOVED***
      return resolve(browserslist.defaults, context)
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^dead$/i,
    select: function (context) ***REMOVED***
      var dead = [
        'Baidu >= 0',
        'ie <= 10',
        'ie_mob <= 11',
        'bb <= 10',
        'op_mob <= 12.1',
        'samsung 4'
      ]
      return resolve(dead, context)
***REMOVED***
***REMOVED***
  ***REMOVED***
    regexp: /^(\w+)$/i,
    select: function (context, name) ***REMOVED***
      if (byName(name, context)) ***REMOVED***
        throw new BrowserslistError(
          'Specify versions in Browserslist query for browser ' + name
        )
  ***REMOVED*** else ***REMOVED***
        throw unknownQuery(name)
  ***REMOVED***
***REMOVED***
***REMOVED***
]

// Get and convert Can I Use data

;(function () ***REMOVED***
  for (var name in agents) ***REMOVED***
    var browser = agents[name]
    browserslist.data[name] = ***REMOVED***
      name: name,
      versions: normalize(agents[name].versions),
      released: normalize(agents[name].versions.slice(0, -3)),
      releaseDate: agents[name].release_date
***REMOVED***
    fillUsage(browserslist.usage.global, name, browser.usage_global)

    browserslist.versionAliases[name] = ***REMOVED******REMOVED***
    for (var i = 0; i < browser.versions.length; i++) ***REMOVED***
      var full = browser.versions[i]
      if (!full) continue

      if (full.indexOf('-') !== -1) ***REMOVED***
        var interval = full.split('-')
        for (var j = 0; j < interval.length; j++) ***REMOVED***
          browserslist.versionAliases[name][interval[j]] = full
    ***REMOVED***
  ***REMOVED***
***REMOVED***
***REMOVED***

  browserslist.versionAliases.op_mob['59'] = '58'

  browserslist.nodeVersions = jsReleases.map(function (release) ***REMOVED***
    return release.version
***REMOVED***)
***REMOVED***)()

module.exports = browserslist
