// yaml/dist/test-events.js
'use strict';

var parseCst = require('./parse-cst.js');
var Document = require('./Document-9b4560a1.js');
require('./PlainValue-ec8e588e.js');
require('./resolveSeq-d03cb037.js');
require('./Schema-88e323a7.js');
require('./warnings-1000a372.js');

function testEvents(src, options) ***REMOVED***
  const opt = Object.assign(***REMOVED***
    keepCstNodes: true,
    keepNodeTypes: true,
    version: '1.2'
***REMOVED*** options);
  const docs = parseCst.parse(src).map(cstDoc => new Document.Document(opt).parse(cstDoc));
  const errDoc = docs.find(doc => doc.errors.length > 0);
  const error = errDoc ? errDoc.errors[0].message : null;
  const events = ['+STR'];

  try ***REMOVED***
    for (let i = 0; i < docs.length; ++i) ***REMOVED***
      const doc = docs[i];
      let root = doc.contents;
      if (Array.isArray(root)) root = root[0];
      const [rootStart, rootEnd] = doc.range || [0, 0];
      let e = doc.errors[0] && doc.errors[0].source;
      if (e && e.type === 'SEQ_ITEM') e = e.node;
      if (e && (e.type === 'DOCUMENT' || e.range.start < rootStart)) throw new Error();
      let docStart = '+DOC';
      const pre = src.slice(0, rootStart);
      const explicitDoc = /---\s*$/.test(pre);
      if (explicitDoc) docStart += ' ---';else if (!doc.contents) continue;
      events.push(docStart);
      addEvents(events, doc, e, root);
      if (doc.contents && doc.contents.length > 1) throw new Error();
      let docEnd = '-DOC';

      if (rootEnd) ***REMOVED***
        const post = src.slice(rootEnd);
        if (/^\.\.\./.test(post)) docEnd += ' ...';
  ***REMOVED***

      events.push(docEnd);
***REMOVED***
***REMOVED*** catch (e) ***REMOVED***
    return ***REMOVED***
      events,
      error: error || e
***REMOVED***;
***REMOVED***

  events.push('-STR');
  return ***REMOVED***
    events,
    error
***REMOVED***;
***REMOVED***

function addEvents(events, doc, e, node) ***REMOVED***
  if (!node) ***REMOVED***
    events.push('=VAL :');
    return;
***REMOVED***

  if (e && node.cstNode === e) throw new Error();
  let props = '';
  let anchor = doc.anchors.getName(node);

  if (anchor) ***REMOVED***
    if (/\d$/.test(anchor)) ***REMOVED***
      const alt = anchor.replace(/\d$/, '');
      if (doc.anchors.getNode(alt)) anchor = alt;
***REMOVED***

    props = ` &$***REMOVED***anchor***REMOVED***`;
***REMOVED***

  if (node.cstNode && node.cstNode.tag) ***REMOVED***
    const ***REMOVED***
      handle,
      suffix
***REMOVED*** = node.cstNode.tag;
    props += handle === '!' && !suffix ? ' <!>' : ` <$***REMOVED***node.tag***REMOVED***>`;
***REMOVED***

  let scalar = null;

  switch (node.type) ***REMOVED***
    case 'ALIAS':
      ***REMOVED***
        let alias = doc.anchors.getName(node.source);

        if (/\d$/.test(alias)) ***REMOVED***
          const alt = alias.replace(/\d$/, '');
          if (doc.anchors.getNode(alt)) alias = alt;
    ***REMOVED***

        events.push(`=ALI$***REMOVED***props***REMOVED*** *$***REMOVED***alias***REMOVED***`);
  ***REMOVED***
      break;

    case 'BLOCK_FOLDED':
      scalar = '>';
      break;

    case 'BLOCK_LITERAL':
      scalar = '|';
      break;

    case 'PLAIN':
      scalar = ':';
      break;

    case 'QUOTE_DOUBLE':
      scalar = '"';
      break;

    case 'QUOTE_SINGLE':
      scalar = "'";
      break;

    case 'PAIR':
      events.push(`+MAP$***REMOVED***props***REMOVED***`);
      addEvents(events, doc, e, node.key);
      addEvents(events, doc, e, node.value);
      events.push('-MAP');
      break;

    case 'FLOW_SEQ':
    case 'SEQ':
      events.push(`+SEQ$***REMOVED***props***REMOVED***`);
      node.items.forEach(item => ***REMOVED***
        addEvents(events, doc, e, item);
  ***REMOVED***);
      events.push('-SEQ');
      break;

    case 'FLOW_MAP':
    case 'MAP':
      events.push(`+MAP$***REMOVED***props***REMOVED***`);
      node.items.forEach((***REMOVED***
        key,
        value
  ***REMOVED***) => ***REMOVED***
        addEvents(events, doc, e, key);
        addEvents(events, doc, e, value);
  ***REMOVED***);
      events.push('-MAP');
      break;

    default:
      throw new Error(`Unexpected node type $***REMOVED***node.type***REMOVED***`);
***REMOVED***

  if (scalar) ***REMOVED***
    const value = node.cstNode.strValue.replace(/\\/g, '\\\\').replace(/\0/g, '\\0').replace(/\x07/g, '\\a').replace(/\x08/g, '\\b').replace(/\t/g, '\\t').replace(/\n/g, '\\n').replace(/\v/g, '\\v').replace(/\f/g, '\\f').replace(/\r/g, '\\r').replace(/\x1b/g, '\\e');
    events.push(`=VAL$***REMOVED***props***REMOVED*** $***REMOVED***scalar***REMOVED***$***REMOVED***value***REMOVED***`);
***REMOVED***
***REMOVED***

exports.testEvents = testEvents;
