// fast-glob/out/index.js
"use strict";
const taskManager = require("./managers/tasks");
const patternManager = require("./managers/patterns");
const async_1 = require("./providers/async");
const stream_1 = require("./providers/stream");
const sync_1 = require("./providers/sync");
const settings_1 = require("./settings");
const utils = require("./utils");
async function FastGlob(source, options) ***REMOVED***
    assertPatternsInput(source);
    const works = getWorks(source, async_1.default, options);
    const result = await Promise.all(works);
    return utils.array.flatten(result);
***REMOVED***
// https://github.com/typescript-eslint/typescript-eslint/issues/60
// eslint-disable-next-line no-redeclare
(function (FastGlob) ***REMOVED***
    function sync(source, options) ***REMOVED***
        assertPatternsInput(source);
        const works = getWorks(source, sync_1.default, options);
        return utils.array.flatten(works);
***REMOVED***
    FastGlob.sync = sync;
    function stream(source, options) ***REMOVED***
        assertPatternsInput(source);
        const works = getWorks(source, stream_1.default, options);
        /**
         * The stream returned by the provider cannot work with an asynchronous iterator.
         * To support asynchronous iterators, regardless of the number of tasks, we always multiplex streams.
         * This affects performance (+25%). I don't see best solution right now.
         */
        return utils.stream.merge(works);
***REMOVED***
    FastGlob.stream = stream;
    function generateTasks(source, options) ***REMOVED***
        assertPatternsInput(source);
        const patterns = patternManager.transform([].concat(source));
        const settings = new settings_1.default(options);
        return taskManager.generate(patterns, settings);
***REMOVED***
    FastGlob.generateTasks = generateTasks;
    function isDynamicPattern(source, options) ***REMOVED***
        assertPatternsInput(source);
        const settings = new settings_1.default(options);
        return utils.pattern.isDynamicPattern(source, settings);
***REMOVED***
    FastGlob.isDynamicPattern = isDynamicPattern;
    function escapePath(source) ***REMOVED***
        assertPatternsInput(source);
        return utils.path.escape(source);
***REMOVED***
    FastGlob.escapePath = escapePath;
***REMOVED***)(FastGlob || (FastGlob = ***REMOVED******REMOVED***));
function getWorks(source, _Provider, options) ***REMOVED***
    const patterns = patternManager.transform([].concat(source));
    const settings = new settings_1.default(options);
    const tasks = taskManager.generate(patterns, settings);
    const provider = new _Provider(settings);
    return tasks.map(provider.read, provider);
***REMOVED***
function assertPatternsInput(input) ***REMOVED***
    const source = [].concat(input);
    const isValidSource = source.every((item) => utils.string.isString(item) && !utils.string.isEmpty(item));
    if (!isValidSource) ***REMOVED***
        throw new TypeError('Patterns must be a string (non empty) or an array of strings');
***REMOVED***
***REMOVED***
module.exports = FastGlob;
