// fraction.js/bigfraction.js
/**
 * @license Fraction.js v4.2.0 23/05/2021
 * https://www.xarg.org/2014/03/rational-numbers-in-javascript/
 *
 * Copyright (c) 2021, Robert Eisele (robert@xarg.org)
 * Dual licensed under the MIT or GPL Version 2 licenses.
 **/


/**
 *
 * This class offers the possibility to calculate fractions.
 * You can pass a fraction in different formats. Either as array, as double, as string or as an integer.
 *
 * Array/Object form
 * [ 0 => <nominator>, 1 => <denominator> ]
 * [ n => <nominator>, d => <denominator> ]
 *
 * Integer form
 * - Single integer value
 *
 * Double form
 * - Single double value
 *
 * String form
 * 123.456 - a simple double
 * 123/456 - a string fraction
 * 123.'456' - a double with repeating decimal places
 * 123.(456) - synonym
 * 123.45'6' - a double with repeating last place
 * 123.45(6) - synonym
 *
 * Example:
 *
 * let f = new Fraction("9.4'31'");
 * f.mul([-4, 3]).div(4.9);
 *
 */

(function(root) ***REMOVED***

  "use strict";

  // Set Identity function to downgrade BigInt to Number if needed
  if (!BigInt) BigInt = function(n) ***REMOVED*** if (isNaN(n)) throw new Error(""); return n; ***REMOVED***;

  const C_ONE = BigInt(1);
  const C_ZERO = BigInt(0);
  const C_TEN = BigInt(10);
  const C_TWO = BigInt(2);
  const C_FIVE = BigInt(5);

  // Maximum search depth for cyclic rational numbers. 2000 should be more than enough.
  // Example: 1/7 = 0.(142857) has 6 repeating decimal places.
  // If MAX_CYCLE_LEN gets reduced, long cycles will not be detected and toString() only gets the first 10 digits
  const MAX_CYCLE_LEN = 2000;

  // Parsed data to avoid calling "new" all the time
  const P = ***REMOVED***
    "s": C_ONE,
    "n": C_ZERO,
    "d": C_ONE
***REMOVED***;

  function assign(n, s) ***REMOVED***

    try ***REMOVED***
      n = BigInt(n);
***REMOVED*** catch (e) ***REMOVED***
      throw Fraction['InvalidParameter'];
***REMOVED***
    return n * s;
***REMOVED***

  // Creates a new Fraction internally without the need of the bulky constructor
  function newFraction(n, d) ***REMOVED***

    if (d === C_ZERO) ***REMOVED***
      throw Fraction['DivisionByZero'];
***REMOVED***

    const f = Object.create(Fraction.prototype);
    f["s"] = n < C_ZERO ? -C_ONE : C_ONE;

    n = n < C_ZERO ? -n : n;

    const a = gcd(n, d);

    f["n"] = n / a;
    f["d"] = d / a;
    return f;
***REMOVED***

  function factorize(num) ***REMOVED***

    const factors = ***REMOVED******REMOVED***;

    let n = num;
    let i = C_TWO;
    let s = C_FIVE - C_ONE;

    while (s <= n) ***REMOVED***

      while (n % i === C_ZERO) ***REMOVED***
        n/= i;
        factors[i] = (factors[i] || C_ZERO) + C_ONE;
  ***REMOVED***
      s+= C_ONE + C_TWO * i++;
***REMOVED***

    if (n !== num) ***REMOVED***
      if (n > 1)
        factors[n] = (factors[n] || C_ZERO) + C_ONE;
***REMOVED*** else ***REMOVED***
      factors[num] = (factors[num] || C_ZERO) + C_ONE;
***REMOVED***
    return factors;
***REMOVED***

  const parse = function(p1, p2) ***REMOVED***

    let n = C_ZERO, d = C_ONE, s = C_ONE;

    if (p1 === undefined || p1 === null) ***REMOVED***
      /* void */
***REMOVED*** else if (p2 !== undefined) ***REMOVED***
      n = BigInt(p1);
      d = BigInt(p2);
      s = n * d;

      if (n % C_ONE !== C_ZERO || d % C_ONE !== C_ZERO) ***REMOVED***
        throw Fraction['NonIntegerParameter'];
  ***REMOVED***

***REMOVED*** else if (typeof p1 === "object") ***REMOVED***
      if ("d" in p1 && "n" in p1) ***REMOVED***
        n = BigInt(p1["n"]);
        d = BigInt(p1["d"]);
        if ("s" in p1)
          n*= BigInt(p1["s"]);
  ***REMOVED*** else if (0 in p1) ***REMOVED***
        n = BigInt(p1[0]);
        if (1 in p1)
          d = BigInt(p1[1]);
  ***REMOVED*** else if (p1 instanceof BigInt) ***REMOVED***
        n = BigInt(p1);
  ***REMOVED*** else ***REMOVED***
        throw Fraction['InvalidParameter'];
  ***REMOVED***
      s = n * d;
***REMOVED*** else if (typeof p1 === "bigint") ***REMOVED***
      n = p1;
      s = p1;
      d = BigInt(1);
***REMOVED*** else if (typeof p1 === "number") ***REMOVED***

      if (isNaN(p1)) ***REMOVED***
        throw Fraction['InvalidParameter'];
  ***REMOVED***

      if (p1 < 0) ***REMOVED***
        s = -C_ONE;
        p1 = -p1;
  ***REMOVED***

      if (p1 % 1 === 0) ***REMOVED***
        n = BigInt(p1);
  ***REMOVED*** else if (p1 > 0) ***REMOVED*** // check for != 0, scale would become NaN (log(0)), which converges really slow

        let z = 1;

        let A = 0, B = 1;
        let C = 1, D = 1;

        let N = 10000000;

        if (p1 >= 1) ***REMOVED***
          z = 10 ** Math.floor(1 + Math.log10(p1));
          p1/= z;
    ***REMOVED***

        // Using Farey Sequences

        while (B <= N && D <= N) ***REMOVED***
          let M = (A + C) / (B + D);

          if (p1 === M) ***REMOVED***
            if (B + D <= N) ***REMOVED***
              n = A + C;
              d = B + D;
        ***REMOVED*** else if (D > B) ***REMOVED***
              n = C;
              d = D;
        ***REMOVED*** else ***REMOVED***
              n = A;
              d = B;
        ***REMOVED***
            break;

      ***REMOVED*** else ***REMOVED***

            if (p1 > M) ***REMOVED***
              A+= C;
              B+= D;
        ***REMOVED*** else ***REMOVED***
              C+= A;
              D+= B;
        ***REMOVED***

            if (B > N) ***REMOVED***
              n = C;
              d = D;
        ***REMOVED*** else ***REMOVED***
              n = A;
              d = B;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
        n = BigInt(n) * BigInt(z);
        d = BigInt(d);

  ***REMOVED***

***REMOVED*** else if (typeof p1 === "string") ***REMOVED***

      let ndx = 0;

      let v = C_ZERO, w = C_ZERO, x = C_ZERO, y = C_ONE, z = C_ONE;

      let match = p1.match(/\d+|./g);

      if (match === null)
        throw Fraction['InvalidParameter'];

      if (match[ndx] === '-') ***REMOVED***// Check for minus sign at the beginning
        s = -C_ONE;
        ndx++;
  ***REMOVED*** else if (match[ndx] === '+') ***REMOVED***// Check for plus sign at the beginning
        ndx++;
  ***REMOVED***

      if (match.length === ndx + 1) ***REMOVED*** // Check if it's just a simple number "1234"
        w = assign(match[ndx++], s);
  ***REMOVED*** else if (match[ndx + 1] === '.' || match[ndx] === '.') ***REMOVED*** // Check if it's a decimal number

        if (match[ndx] !== '.') ***REMOVED*** // Handle 0.5 and .5
          v = assign(match[ndx++], s);
    ***REMOVED***
        ndx++;

        // Check for decimal places
        if (ndx + 1 === match.length || match[ndx + 1] === '(' && match[ndx + 3] === ')' || match[ndx + 1] === "'" && match[ndx + 3] === "'") ***REMOVED***
          w = assign(match[ndx], s);
          y = C_TEN ** BigInt(match[ndx].length);
          ndx++;
    ***REMOVED***

        // Check for repeating places
        if (match[ndx] === '(' && match[ndx + 2] === ')' || match[ndx] === "'" && match[ndx + 2] === "'") ***REMOVED***
          x = assign(match[ndx + 1], s);
          z = C_TEN ** BigInt(match[ndx + 1].length) - C_ONE;
          ndx+= 3;
    ***REMOVED***

  ***REMOVED*** else if (match[ndx + 1] === '/' || match[ndx + 1] === ':') ***REMOVED*** // Check for a simple fraction "123/456" or "123:456"
        w = assign(match[ndx], s);
        y = assign(match[ndx + 2], C_ONE);
        ndx+= 3;
  ***REMOVED*** else if (match[ndx + 3] === '/' && match[ndx + 1] === ' ') ***REMOVED*** // Check for a complex fraction "123 1/2"
        v = assign(match[ndx], s);
        w = assign(match[ndx + 2], s);
        y = assign(match[ndx + 4], C_ONE);
        ndx+= 5;
  ***REMOVED***

      if (match.length <= ndx) ***REMOVED*** // Check for more tokens on the stack
        d = y * z;
        s = /* void */
        n = x + d * v + z * w;
  ***REMOVED*** else ***REMOVED***
        throw Fraction['InvalidParameter'];
  ***REMOVED***

***REMOVED*** else ***REMOVED***
      throw Fraction['InvalidParameter'];
***REMOVED***

    if (d === C_ZERO) ***REMOVED***
      throw Fraction['DivisionByZero'];
***REMOVED***

    P["s"] = s < C_ZERO ? -C_ONE : C_ONE;
    P["n"] = n < C_ZERO ? -n : n;
    P["d"] = d < C_ZERO ? -d : d;
***REMOVED***;

  function modpow(b, e, m) ***REMOVED***

    let r = C_ONE;
    for (; e > C_ZERO; b = (b * b) % m, e >>= C_ONE) ***REMOVED***

      if (e & C_ONE) ***REMOVED***
        r = (r * b) % m;
  ***REMOVED***
***REMOVED***
    return r;
***REMOVED***

  function cycleLen(n, d) ***REMOVED***

    for (; d % C_TWO === C_ZERO;
      d/= C_TWO) ***REMOVED***
***REMOVED***

    for (; d % C_FIVE === C_ZERO;
      d/= C_FIVE) ***REMOVED***
***REMOVED***

    if (d === C_ONE) // Catch non-cyclic numbers
      return C_ZERO;

    // If we would like to compute really large numbers quicker, we could make use of Fermat's little theorem:
    // 10^(d-1) % d == 1
    // However, we don't need such large numbers and MAX_CYCLE_LEN should be the capstone,
    // as we want to translate the numbers to strings.

    let rem = C_TEN % d;
    let t = 1;

    for (; rem !== C_ONE; t++) ***REMOVED***
      rem = rem * C_TEN % d;

      if (t > MAX_CYCLE_LEN)
        return C_ZERO; // Returning 0 here means that we don't print it as a cyclic number. It's likely that the answer is `d-1`
***REMOVED***
    return BigInt(t);
***REMOVED***

  function cycleStart(n, d, len) ***REMOVED***

    let rem1 = C_ONE;
    let rem2 = modpow(C_TEN, len, d);

    for (let t = 0; t < 300; t++) ***REMOVED*** // s < ~log10(Number.MAX_VALUE)
      // Solve 10^s == 10^(s+t) (mod d)

      if (rem1 === rem2)
        return BigInt(t);

      rem1 = rem1 * C_TEN % d;
      rem2 = rem2 * C_TEN % d;
***REMOVED***
    return 0;
***REMOVED***

  function gcd(a, b) ***REMOVED***

    if (!a)
      return b;
    if (!b)
      return a;

    while (1) ***REMOVED***
      a%= b;
      if (!a)
        return b;
      b%= a;
      if (!b)
        return a;
***REMOVED***
***REMOVED***

  /**
   * Module constructor
   *
   * @constructor
   * @param ***REMOVED***number|Fraction=***REMOVED*** a
   * @param ***REMOVED***number=***REMOVED*** b
   */
  function Fraction(a, b) ***REMOVED***

    parse(a, b);

    if (this instanceof Fraction) ***REMOVED***
      a = gcd(P["d"], P["n"]); // Abuse a
      this["s"] = P["s"];
      this["n"] = P["n"] / a;
      this["d"] = P["d"] / a;
***REMOVED*** else ***REMOVED***
      return newFraction(P['s'] * P['n'], P['d']);
***REMOVED***
***REMOVED***

  Fraction['DivisionByZero'] = new Error("Division by Zero");
  Fraction['InvalidParameter'] = new Error("Invalid argument");
  Fraction['NonIntegerParameter'] = new Error("Parameters must be integer");

  Fraction.prototype = ***REMOVED***

    "s": C_ONE,
    "n": C_ZERO,
    "d": C_ONE,

    /**
     * Calculates the absolute value
     *
     * Ex: new Fraction(-4).abs() => 4
     **/
    "abs": function() ***REMOVED***

      return newFraction(this["n"], this["d"]);
***REMOVED***,

    /**
     * Inverts the sign of the current fraction
     *
     * Ex: new Fraction(-4).neg() => 4
     **/
    "neg": function() ***REMOVED***

      return newFraction(-this["s"] * this["n"], this["d"]);
***REMOVED***,

    /**
     * Adds two rational numbers
     *
     * Ex: new Fraction(***REMOVED***n: 2, d: 3***REMOVED***).add("14.9") => 467 / 30
     **/
    "add": function(a, b) ***REMOVED***

      parse(a, b);
      return newFraction(
        this["s"] * this["n"] * P["d"] + P["s"] * this["d"] * P["n"],
        this["d"] * P["d"]
      );
***REMOVED***,

    /**
     * Subtracts two rational numbers
     *
     * Ex: new Fraction(***REMOVED***n: 2, d: 3***REMOVED***).add("14.9") => -427 / 30
     **/
    "sub": function(a, b) ***REMOVED***

      parse(a, b);
      return newFraction(
        this["s"] * this["n"] * P["d"] - P["s"] * this["d"] * P["n"],
        this["d"] * P["d"]
      );
***REMOVED***,

    /**
     * Multiplies two rational numbers
     *
     * Ex: new Fraction("-17.(345)").mul(3) => 5776 / 111
     **/
    "mul": function(a, b) ***REMOVED***

      parse(a, b);
      return newFraction(
        this["s"] * P["s"] * this["n"] * P["n"],
        this["d"] * P["d"]
      );
***REMOVED***,

    /**
     * Divides two rational numbers
     *
     * Ex: new Fraction("-17.(345)").inverse().div(3)
     **/
    "div": function(a, b) ***REMOVED***

      parse(a, b);
      return newFraction(
        this["s"] * P["s"] * this["n"] * P["d"],
        this["d"] * P["n"]
      );
***REMOVED***,

    /**
     * Clones the actual object
     *
     * Ex: new Fraction("-17.(345)").clone()
     **/
    "clone": function() ***REMOVED***
      return newFraction(this['s'] * this['n'], this['d']);
***REMOVED***,

    /**
     * Calculates the modulo of two rational numbers - a more precise fmod
     *
     * Ex: new Fraction('4.(3)').mod([7, 8]) => (13/3) % (7/8) = (5/6)
     **/
    "mod": function(a, b) ***REMOVED***

      if (a === undefined) ***REMOVED***
        return newFraction(this["s"] * this["n"] % this["d"], C_ONE);
  ***REMOVED***

      parse(a, b);
      if (0 === P["n"] && 0 === this["d"]) ***REMOVED***
        throw Fraction['DivisionByZero'];
  ***REMOVED***

      /*
       * First silly attempt, kinda slow
       *
       return that["sub"](***REMOVED***
       "n": num["n"] * Math.floor((this.n / this.d) / (num.n / num.d)),
       "d": num["d"],
       "s": this["s"]
   ***REMOVED***);*/

      /*
       * New attempt: a1 / b1 = a2 / b2 * q + r
       * => b2 * a1 = a2 * b1 * q + b1 * b2 * r
       * => (b2 * a1 % a2 * b1) / (b1 * b2)
       */
      return newFraction(
        this["s"] * (P["d"] * this["n"]) % (P["n"] * this["d"]),
        P["d"] * this["d"]
      );
***REMOVED***,

    /**
     * Calculates the fractional gcd of two rational numbers
     *
     * Ex: new Fraction(5,8).gcd(3,7) => 1/56
     */
    "gcd": function(a, b) ***REMOVED***

      parse(a, b);

      // gcd(a / b, c / d) = gcd(a, c) / lcm(b, d)

      return newFraction(gcd(P["n"], this["n"]) * gcd(P["d"], this["d"]), P["d"] * this["d"]);
***REMOVED***,

    /**
     * Calculates the fractional lcm of two rational numbers
     *
     * Ex: new Fraction(5,8).lcm(3,7) => 15
     */
    "lcm": function(a, b) ***REMOVED***

      parse(a, b);

      // lcm(a / b, c / d) = lcm(a, c) / gcd(b, d)

      if (P["n"] === C_ZERO && this["n"] === C_ZERO) ***REMOVED***
        return newFraction(C_ZERO, C_ONE);
  ***REMOVED***
      return newFraction(P["n"] * this["n"], gcd(P["n"], this["n"]) * gcd(P["d"], this["d"]));
***REMOVED***,

    /**
     * Gets the inverse of the fraction, means numerator and denominator are exchanged
     *
     * Ex: new Fraction([-3, 4]).inverse() => -4 / 3
     **/
    "inverse": function() ***REMOVED***
      return newFraction(this["s"] * this["d"], this["n"]);
***REMOVED***,

    /**
     * Calculates the fraction to some integer exponent
     *
     * Ex: new Fraction(-1,2).pow(-3) => -8
     */
    "pow": function(a, b) ***REMOVED***

      parse(a, b);

      // Trivial case when exp is an integer

      if (P['d'] === C_ONE) ***REMOVED***

        if (P['s'] < C_ZERO) ***REMOVED***
          return newFraction((this['s'] * this["d"]) ** P['n'], this["n"] ** P['n']);
    ***REMOVED*** else ***REMOVED***
          return newFraction((this['s'] * this["n"]) ** P['n'], this["d"] ** P['n']);
    ***REMOVED***
  ***REMOVED***

      // Negative roots become complex
      //     (-a/b)^(c/d) = x
      // <=> (-1)^(c/d) * (a/b)^(c/d) = x
      // <=> (cos(pi) + i*sin(pi))^(c/d) * (a/b)^(c/d) = x
      // <=> (cos(c*pi/d) + i*sin(c*pi/d)) * (a/b)^(c/d) = x       # DeMoivre's formula
      // From which follows that only for c=0 the root is non-complex
      if (this['s'] < C_ZERO) return null;

      // Now prime factor n and d
      let N = factorize(this['n']);
      let D = factorize(this['d']);

      // Exponentiate and take root for n and d individually
      let n = C_ONE;
      let d = C_ONE;
      for (let k in N) ***REMOVED***
        if (k === '1') continue;
        if (k === '0') ***REMOVED***
          n = C_ZERO;
          break;
    ***REMOVED***
        N[k]*= P['n'];

        if (N[k] % P['d'] === C_ZERO) ***REMOVED***
          N[k]/= P['d'];
    ***REMOVED*** else return null;
        n*= BigInt(k) ** N[k];
  ***REMOVED***

      for (let k in D) ***REMOVED***
        if (k === '1') continue;
        D[k]*= P['n'];

        if (D[k] % P['d'] === C_ZERO) ***REMOVED***
          D[k]/= P['d'];
    ***REMOVED*** else return null;
        d*= BigInt(k) ** D[k];
  ***REMOVED***

      if (P['s'] < C_ZERO) ***REMOVED***
        return newFraction(d, n);
  ***REMOVED***
      return newFraction(n, d);
***REMOVED***,

    /**
     * Check if two rational numbers are the same
     *
     * Ex: new Fraction(19.6).equals([98, 5]);
     **/
    "equals": function(a, b) ***REMOVED***

      parse(a, b);
      return this["s"] * this["n"] * P["d"] === P["s"] * P["n"] * this["d"]; // Same as compare() === 0
***REMOVED***,

    /**
     * Check if two rational numbers are the same
     *
     * Ex: new Fraction(19.6).equals([98, 5]);
     **/
    "compare": function(a, b) ***REMOVED***

      parse(a, b);
      let t = (this["s"] * this["n"] * P["d"] - P["s"] * P["n"] * this["d"]);

      return (C_ZERO < t) - (t < C_ZERO);
***REMOVED***,

    /**
     * Calculates the ceil of a rational number
     *
     * Ex: new Fraction('4.(3)').ceil() => (5 / 1)
     **/
    "ceil": function(places) ***REMOVED***

      places = C_TEN ** BigInt(places || 0);

      return newFraction(this["s"] * places * this["n"] / this["d"] +
        (places * this["n"] % this["d"] > C_ZERO && this["s"] >= C_ZERO ? C_ONE : C_ZERO),
        places);
***REMOVED***,

    /**
     * Calculates the floor of a rational number
     *
     * Ex: new Fraction('4.(3)').floor() => (4 / 1)
     **/
    "floor": function(places) ***REMOVED***

      places = C_TEN ** BigInt(places || 0);

      return newFraction(this["s"] * places * this["n"] / this["d"] -
        (places * this["n"] % this["d"] > C_ZERO && this["s"] < C_ZERO ? C_ONE : C_ZERO),
        places);
***REMOVED***,

    /**
     * Rounds a rational numbers
     *
     * Ex: new Fraction('4.(3)').round() => (4 / 1)
     **/
    "round": function(places) ***REMOVED***

      places = C_TEN ** BigInt(places || 0);

      /* Derivation:

      s >= 0:
        round(n / d) = trunc(n / d) + (n % d) / d >= 0.5 ? 1 : 0
                     = trunc(n / d) + 2(n % d) >= d ? 1 : 0
      s < 0:
        round(n / d) =-trunc(n / d) - (n % d) / d > 0.5 ? 1 : 0
                     =-trunc(n / d) - 2(n % d) > d ? 1 : 0

      =>:

      round(s * n / d) = s * trunc(n / d) + s * (C + 2(n % d) > d ? 1 : 0)
          where C = s >= 0 ? 1 : 0, to fix the >= for the positve case.
      */

      return newFraction(this["s"] * places * this["n"] / this["d"] +
        this["s"] * ((this["s"] >= C_ZERO ? C_ONE : C_ZERO) + C_TWO * (places * this["n"] % this["d"]) > this["d"] ? C_ONE : C_ZERO),
        places);
***REMOVED***,

    /**
     * Check if two rational numbers are divisible
     *
     * Ex: new Fraction(19.6).divisible(1.5);
     */
    "divisible": function(a, b) ***REMOVED***

      parse(a, b);
      return !(!(P["n"] * this["d"]) || ((this["n"] * P["d"]) % (P["n"] * this["d"])));
***REMOVED***,

    /**
     * Returns a decimal representation of the fraction
     *
     * Ex: new Fraction("100.'91823'").valueOf() => 100.91823918239183
     **/
    'valueOf': function() ***REMOVED***
      // Best we can do so far
      return Number(this["s"] * this["n"]) / Number(this["d"]);
***REMOVED***,

    /**
     * Creates a string representation of a fraction with all digits
     *
     * Ex: new Fraction("100.'91823'").toString() => "100.(91823)"
     **/
    'toString': function(dec) ***REMOVED***

      let N = this["n"];
      let D = this["d"];

      dec = dec || 15; // 15 = decimal places when no repitation

      let cycLen = cycleLen(N, D); // Cycle length
      let cycOff = cycleStart(N, D, cycLen); // Cycle start

      let str = this['s'] < C_ZERO ? "-" : "";

      // Append integer part
      str+= N / D;

      N%= D;
      N*= C_TEN;

      if (N)
        str+= ".";

      if (cycLen) ***REMOVED***

        for (let i = cycOff; i--;) ***REMOVED***
          str+= N / D;
          N%= D;
          N*= C_TEN;
    ***REMOVED***
        str+= "(";
        for (let i = cycLen; i--;) ***REMOVED***
          str+= N / D;
          N%= D;
          N*= C_TEN;
    ***REMOVED***
        str+= ")";
  ***REMOVED*** else ***REMOVED***
        for (let i = dec; N && i--;) ***REMOVED***
          str+= N / D;
          N%= D;
          N*= C_TEN;
    ***REMOVED***
  ***REMOVED***
      return str;
***REMOVED***,

    /**
     * Returns a string-fraction representation of a Fraction object
     *
     * Ex: new Fraction("1.'3'").toFraction() => "4 1/3"
     **/
    'toFraction': function(excludeWhole) ***REMOVED***

      let n = this["n"];
      let d = this["d"];
      let str = this['s'] < C_ZERO ? "-" : "";

      if (d === C_ONE) ***REMOVED***
        str+= n;
  ***REMOVED*** else ***REMOVED***
        let whole = n / d;
        if (excludeWhole && whole > C_ZERO) ***REMOVED***
          str+= whole;
          str+= " ";
          n%= d;
    ***REMOVED***

        str+= n;
        str+= '/';
        str+= d;
  ***REMOVED***
      return str;
***REMOVED***,

    /**
     * Returns a latex representation of a Fraction object
     *
     * Ex: new Fraction("1.'3'").toLatex() => "\frac***REMOVED***4***REMOVED******REMOVED***3***REMOVED***"
     **/
    'toLatex': function(excludeWhole) ***REMOVED***

      let n = this["n"];
      let d = this["d"];
      let str = this['s'] < C_ZERO ? "-" : "";

      if (d === C_ONE) ***REMOVED***
        str+= n;
  ***REMOVED*** else ***REMOVED***
        let whole = n / d;
        if (excludeWhole && whole > C_ZERO) ***REMOVED***
          str+= whole;
          n%= d;
    ***REMOVED***

        str+= "\\frac***REMOVED***";
        str+= n;
        str+= '***REMOVED******REMOVED***';
        str+= d;
        str+= '***REMOVED***';
  ***REMOVED***
      return str;
***REMOVED***,

    /**
     * Returns an array of continued fraction elements
     *
     * Ex: new Fraction("7/8").toContinued() => [0,1,7]
     */
    'toContinued': function() ***REMOVED***

      let a = this['n'];
      let b = this['d'];
      let res = [];

      do ***REMOVED***
        res.push(a / b);
        let t = a % b;
        a = b;
        b = t;
  ***REMOVED*** while (a !== C_ONE);

      return res;
***REMOVED***,

    "simplify": function(eps) ***REMOVED***

      eps = eps || 0.001;

      const thisABS = this['abs']();
      const cont = thisABS['toContinued']();

      for (let i = 1; i < cont.length; i++) ***REMOVED***

        let s = newFraction(cont[i - 1], C_ONE);
        for (let k = i - 2; k >= 0; k--) ***REMOVED***
          s = s['inverse']()['add'](cont[k]);
    ***REMOVED***

        if (s['sub'](thisABS)['abs']().valueOf() < eps) ***REMOVED***
          return s['mul'](this['s']);
    ***REMOVED***
  ***REMOVED***
      return this;
***REMOVED***
***REMOVED***;

  if (typeof define === "function" && define["amd"]) ***REMOVED***
    define([], function() ***REMOVED***
      return Fraction;
***REMOVED***);
***REMOVED*** else if (typeof exports === "object") ***REMOVED***
    Object.defineProperty(exports, "__esModule", ***REMOVED*** 'value': true ***REMOVED***);
    Fraction['default'] = Fraction;
    Fraction['Fraction'] = Fraction;
    module['exports'] = Fraction;
***REMOVED*** else ***REMOVED***
    root['Fraction'] = Fraction;
***REMOVED***

***REMOVED***)(this);
