// yargs/build/lib/middleware.js
import ***REMOVED*** argsert ***REMOVED*** from './argsert.js';
import ***REMOVED*** isPromise ***REMOVED*** from './utils/is-promise.js';
export class GlobalMiddleware ***REMOVED***
    constructor(yargs) ***REMOVED***
        this.globalMiddleware = [];
        this.frozens = [];
        this.yargs = yargs;
***REMOVED***
    addMiddleware(callback, applyBeforeValidation, global = true, mutates = false) ***REMOVED***
        argsert('<array|function> [boolean] [boolean] [boolean]', [callback, applyBeforeValidation, global], arguments.length);
        if (Array.isArray(callback)) ***REMOVED***
            for (let i = 0; i < callback.length; i++) ***REMOVED***
                if (typeof callback[i] !== 'function') ***REMOVED***
                    throw Error('middleware must be a function');
            ***REMOVED***
                const m = callback[i];
                m.applyBeforeValidation = applyBeforeValidation;
                m.global = global;
        ***REMOVED***
            Array.prototype.push.apply(this.globalMiddleware, callback);
    ***REMOVED***
        else if (typeof callback === 'function') ***REMOVED***
            const m = callback;
            m.applyBeforeValidation = applyBeforeValidation;
            m.global = global;
            m.mutates = mutates;
            this.globalMiddleware.push(callback);
    ***REMOVED***
        return this.yargs;
***REMOVED***
    addCoerceMiddleware(callback, option) ***REMOVED***
        const aliases = this.yargs.getAliases();
        this.globalMiddleware = this.globalMiddleware.filter(m => ***REMOVED***
            const toCheck = [...(aliases[option] || []), option];
            if (!m.option)
                return true;
            else
                return !toCheck.includes(m.option);
    ***REMOVED***);
        callback.option = option;
        return this.addMiddleware(callback, true, true, true);
***REMOVED***
    getMiddleware() ***REMOVED***
        return this.globalMiddleware;
***REMOVED***
    freeze() ***REMOVED***
        this.frozens.push([...this.globalMiddleware]);
***REMOVED***
    unfreeze() ***REMOVED***
        const frozen = this.frozens.pop();
        if (frozen !== undefined)
            this.globalMiddleware = frozen;
***REMOVED***
    reset() ***REMOVED***
        this.globalMiddleware = this.globalMiddleware.filter(m => m.global);
***REMOVED***
***REMOVED***
export function commandMiddlewareFactory(commandMiddleware) ***REMOVED***
    if (!commandMiddleware)
        return [];
    return commandMiddleware.map(middleware => ***REMOVED***
        middleware.applyBeforeValidation = false;
        return middleware;
***REMOVED***);
***REMOVED***
export function applyMiddleware(argv, yargs, middlewares, beforeValidation) ***REMOVED***
    return middlewares.reduce((acc, middleware) => ***REMOVED***
        if (middleware.applyBeforeValidation !== beforeValidation) ***REMOVED***
            return acc;
    ***REMOVED***
        if (middleware.mutates) ***REMOVED***
            if (middleware.applied)
                return acc;
            middleware.applied = true;
    ***REMOVED***
        if (isPromise(acc)) ***REMOVED***
            return acc
                .then(initialObj => Promise.all([initialObj, middleware(initialObj, yargs)]))
                .then(([initialObj, middlewareObj]) => Object.assign(initialObj, middlewareObj));
    ***REMOVED***
        else ***REMOVED***
            const result = middleware(acc, yargs);
            return isPromise(result)
                ? result.then(middlewareObj => Object.assign(acc, middlewareObj))
                : Object.assign(acc, result);
    ***REMOVED***
***REMOVED***, argv);
***REMOVED***
