// autoprefixer/lib/autoprefixer.js
let browserslist = require('browserslist')
let ***REMOVED*** agents ***REMOVED*** = require('caniuse-lite')
let pico = require('picocolors')

let Browsers = require('./browsers')
let Prefixes = require('./prefixes')
let dataPrefixes = require('../data/prefixes')
let getInfo = require('./info')

let autoprefixerData = ***REMOVED*** browsers: agents, prefixes: dataPrefixes ***REMOVED***

const WARNING =
  '\n' +
  '  Replace Autoprefixer `browsers` option to Browserslist config.\n' +
  '  Use `browserslist` key in `package.json` or `.browserslistrc` file.\n' +
  '\n' +
  '  Using `browsers` option can cause errors. Browserslist config can\n' +
  '  be used for Babel, Autoprefixer, postcss-normalize and other tools.\n' +
  '\n' +
  '  If you really need to use option, rename it to `overrideBrowserslist`.\n' +
  '\n' +
  '  Learn more at:\n' +
  '  https://github.com/browserslist/browserslist#readme\n' +
  '  https://twitter.com/browserslist\n' +
  '\n'

function isPlainObject(obj) ***REMOVED***
  return Object.prototype.toString.apply(obj) === '[object Object]'
***REMOVED***

let cache = new Map()

function timeCapsule(result, prefixes) ***REMOVED***
  if (prefixes.browsers.selected.length === 0) ***REMOVED***
    return
***REMOVED***
  if (prefixes.add.selectors.length > 0) ***REMOVED***
    return
***REMOVED***
  if (Object.keys(prefixes.add).length > 2) ***REMOVED***
    return
***REMOVED***
  /* c8 ignore next 11 */
  result.warn(
    'Autoprefixer target browsers do not need any prefixes.' +
      'You do not need Autoprefixer anymore.\n' +
      'Check your Browserslist config to be sure that your targets ' +
      'are set up correctly.\n' +
      '\n' +
      '  Learn more at:\n' +
      '  https://github.com/postcss/autoprefixer#readme\n' +
      '  https://github.com/browserslist/browserslist#readme\n' +
      '\n'
  )
***REMOVED***

module.exports = plugin

function plugin(...reqs) ***REMOVED***
  let options
  if (reqs.length === 1 && isPlainObject(reqs[0])) ***REMOVED***
    options = reqs[0]
    reqs = undefined
***REMOVED*** else if (reqs.length === 0 || (reqs.length === 1 && !reqs[0])) ***REMOVED***
    reqs = undefined
***REMOVED*** else if (reqs.length <= 2 && (Array.isArray(reqs[0]) || !reqs[0])) ***REMOVED***
    options = reqs[1]
    reqs = reqs[0]
***REMOVED*** else if (typeof reqs[reqs.length - 1] === 'object') ***REMOVED***
    options = reqs.pop()
***REMOVED***

  if (!options) ***REMOVED***
    options = ***REMOVED******REMOVED***
***REMOVED***

  if (options.browser) ***REMOVED***
    throw new Error(
      'Change `browser` option to `overrideBrowserslist` in Autoprefixer'
    )
***REMOVED*** else if (options.browserslist) ***REMOVED***
    throw new Error(
      'Change `browserslist` option to `overrideBrowserslist` in Autoprefixer'
    )
***REMOVED***

  if (options.overrideBrowserslist) ***REMOVED***
    reqs = options.overrideBrowserslist
***REMOVED*** else if (options.browsers) ***REMOVED***
    if (typeof console !== 'undefined' && console.warn) ***REMOVED***
      console.warn(
        pico.red(WARNING.replace(/`[^`]+`/g, i => pico.yellow(i.slice(1, -1))))
      )
***REMOVED***
    reqs = options.browsers
***REMOVED***

  let brwlstOpts = ***REMOVED***
    ignoreUnknownVersions: options.ignoreUnknownVersions,
    stats: options.stats,
    env: options.env
***REMOVED***

  function loadPrefixes(opts) ***REMOVED***
    let d = autoprefixerData
    let browsers = new Browsers(d.browsers, reqs, opts, brwlstOpts)
    let key = browsers.selected.join(', ') + JSON.stringify(options)

    if (!cache.has(key)) ***REMOVED***
      cache.set(key, new Prefixes(d.prefixes, browsers, options))
***REMOVED***

    return cache.get(key)
***REMOVED***

  return ***REMOVED***
    postcssPlugin: 'autoprefixer',

    prepare(result) ***REMOVED***
      let prefixes = loadPrefixes(***REMOVED***
        from: result.opts.from,
        env: options.env
  ***REMOVED***)

      return ***REMOVED***
        OnceExit(root) ***REMOVED***
          timeCapsule(result, prefixes)
          if (options.remove !== false) ***REMOVED***
            prefixes.processor.remove(root, result)
      ***REMOVED***
          if (options.add !== false) ***REMOVED***
            prefixes.processor.add(root, result)
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***,

    info(opts) ***REMOVED***
      opts = opts || ***REMOVED******REMOVED***
      opts.from = opts.from || process.cwd()
      return getInfo(loadPrefixes(opts))
***REMOVED***,

    options,
    browsers: reqs
***REMOVED***
***REMOVED***

plugin.postcss = true

/**
 * Autoprefixer data
 */
plugin.data = autoprefixerData

/**
 * Autoprefixer default browsers
 */
plugin.defaults = browserslist.defaults

/**
 * Inspect with default Autoprefixer
 */
plugin.info = () => plugin().info()
