// autoprefixer/lib/info.js
let browserslist = require('browserslist')

function capitalize(str) ***REMOVED***
  return str.slice(0, 1).toUpperCase() + str.slice(1)
***REMOVED***

const NAMES = ***REMOVED***
  ie: 'IE',
  ie_mob: 'IE Mobile',
  ios_saf: 'iOS Safari',
  op_mini: 'Opera Mini',
  op_mob: 'Opera Mobile',
  and_chr: 'Chrome for Android',
  and_ff: 'Firefox for Android',
  and_uc: 'UC for Android',
  and_qq: 'QQ Browser',
  kaios: 'KaiOS Browser',
  baidu: 'Baidu Browser',
  samsung: 'Samsung Internet'
***REMOVED***

function prefix(name, prefixes, note) ***REMOVED***
  let out = `  $***REMOVED***name***REMOVED***`
  if (note) out += ' *'
  out += ': '
  out += prefixes.map(i => i.replace(/^-(.*)-$/g, '$1')).join(', ')
  out += '\n'
  return out
***REMOVED***

module.exports = function (prefixes) ***REMOVED***
  if (prefixes.browsers.selected.length === 0) ***REMOVED***
    return 'No browsers selected'
***REMOVED***

  let versions = ***REMOVED******REMOVED***
  for (let browser of prefixes.browsers.selected) ***REMOVED***
    let parts = browser.split(' ')
    let name = parts[0]
    let version = parts[1]

    name = NAMES[name] || capitalize(name)
    if (versions[name]) ***REMOVED***
      versions[name].push(version)
***REMOVED*** else ***REMOVED***
      versions[name] = [version]
***REMOVED***
***REMOVED***

  let out = 'Browsers:\n'
  for (let browser in versions) ***REMOVED***
    let list = versions[browser]
    list = list.sort((a, b) => parseFloat(b) - parseFloat(a))
    out += `  $***REMOVED***browser***REMOVED***: $***REMOVED***list.join(', ')***REMOVED***\n`
***REMOVED***

  let coverage = browserslist.coverage(prefixes.browsers.selected)
  let round = Math.round(coverage * 100) / 100.0
  out += `\nThese browsers account for $***REMOVED***round***REMOVED***% of all users globally\n`

  let atrules = []
  for (let name in prefixes.add) ***REMOVED***
    let data = prefixes.add[name]
    if (name[0] === '@' && data.prefixes) ***REMOVED***
      atrules.push(prefix(name, data.prefixes))
***REMOVED***
***REMOVED***
  if (atrules.length > 0) ***REMOVED***
    out += `\nAt-Rules:\n$***REMOVED***atrules.sort().join('')***REMOVED***`
***REMOVED***

  let selectors = []
  for (let selector of prefixes.add.selectors) ***REMOVED***
    if (selector.prefixes) ***REMOVED***
      selectors.push(prefix(selector.name, selector.prefixes))
***REMOVED***
***REMOVED***
  if (selectors.length > 0) ***REMOVED***
    out += `\nSelectors:\n$***REMOVED***selectors.sort().join('')***REMOVED***`
***REMOVED***

  let values = []
  let props = []
  let hadGrid = false
  for (let name in prefixes.add) ***REMOVED***
    let data = prefixes.add[name]
    if (name[0] !== '@' && data.prefixes) ***REMOVED***
      let grid = name.indexOf('grid-') === 0
      if (grid) hadGrid = true
      props.push(prefix(name, data.prefixes, grid))
***REMOVED***

    if (!Array.isArray(data.values)) ***REMOVED***
      continue
***REMOVED***
    for (let value of data.values) ***REMOVED***
      let grid = value.name.includes('grid')
      if (grid) hadGrid = true
      let string = prefix(value.name, value.prefixes, grid)
      if (!values.includes(string)) ***REMOVED***
        values.push(string)
  ***REMOVED***
***REMOVED***
***REMOVED***

  if (props.length > 0) ***REMOVED***
    out += `\nProperties:\n$***REMOVED***props.sort().join('')***REMOVED***`
***REMOVED***
  if (values.length > 0) ***REMOVED***
    out += `\nValues:\n$***REMOVED***values.sort().join('')***REMOVED***`
***REMOVED***
  if (hadGrid) ***REMOVED***
    out += '\n* - Prefixes will be added only on grid: true option.\n'
***REMOVED***

  if (!atrules.length && !selectors.length && !props.length && !values.length) ***REMOVED***
    out +=
      "\nAwesome! Your browsers don't require any vendor prefixes." +
      '\nNow you can remove Autoprefixer from build steps.'
***REMOVED***

  return out
***REMOVED***
