// postcss/lib/tokenize.js
'use strict'

const SINGLE_QUOTE = "'".charCodeAt(0)
const DOUBLE_QUOTE = '"'.charCodeAt(0)
const BACKSLASH = '\\'.charCodeAt(0)
const SLASH = '/'.charCodeAt(0)
const NEWLINE = '\n'.charCodeAt(0)
const SPACE = ' '.charCodeAt(0)
const FEED = '\f'.charCodeAt(0)
const TAB = '\t'.charCodeAt(0)
const CR = '\r'.charCodeAt(0)
const OPEN_SQUARE = '['.charCodeAt(0)
const CLOSE_SQUARE = ']'.charCodeAt(0)
const OPEN_PARENTHESES = '('.charCodeAt(0)
const CLOSE_PARENTHESES = ')'.charCodeAt(0)
const OPEN_CURLY = '***REMOVED***'.charCodeAt(0)
const CLOSE_CURLY = '***REMOVED***'.charCodeAt(0)
const SEMICOLON = ';'.charCodeAt(0)
const ASTERISK = '*'.charCodeAt(0)
const COLON = ':'.charCodeAt(0)
const AT = '@'.charCodeAt(0)

const RE_AT_END = /[\t\n\f\r "#'()/;[\\\]***REMOVED******REMOVED***]/g
const RE_WORD_END = /[\t\n\f\r !"#'():;@[\\\]***REMOVED******REMOVED***]|\/(?=\*)/g
const RE_BAD_BRACKET = /.[\n"'(/\\]/
const RE_HEX_ESCAPE = /[\da-f]/i

module.exports = function tokenizer(input, options = ***REMOVED******REMOVED***) ***REMOVED***
  let css = input.css.valueOf()
  let ignore = options.ignoreErrors

  let code, next, quote, content, escape
  let escaped, escapePos, prev, n, currentToken

  let length = css.length
  let pos = 0
  let buffer = []
  let returned = []

  function position() ***REMOVED***
    return pos
***REMOVED***

  function unclosed(what) ***REMOVED***
    throw input.error('Unclosed ' + what, pos)
***REMOVED***

  function endOfFile() ***REMOVED***
    return returned.length === 0 && pos >= length
***REMOVED***

  function nextToken(opts) ***REMOVED***
    if (returned.length) return returned.pop()
    if (pos >= length) return

    let ignoreUnclosed = opts ? opts.ignoreUnclosed : false

    code = css.charCodeAt(pos)

    switch (code) ***REMOVED***
      case NEWLINE:
      case SPACE:
      case TAB:
      case CR:
      case FEED: ***REMOVED***
        next = pos
        do ***REMOVED***
          next += 1
          code = css.charCodeAt(next)
    ***REMOVED*** while (
          code === SPACE ||
          code === NEWLINE ||
          code === TAB ||
          code === CR ||
          code === FEED
        )

        currentToken = ['space', css.slice(pos, next)]
        pos = next - 1
        break
  ***REMOVED***

      case OPEN_SQUARE:
      case CLOSE_SQUARE:
      case OPEN_CURLY:
      case CLOSE_CURLY:
      case COLON:
      case SEMICOLON:
      case CLOSE_PARENTHESES: ***REMOVED***
        let controlChar = String.fromCharCode(code)
        currentToken = [controlChar, controlChar, pos]
        break
  ***REMOVED***

      case OPEN_PARENTHESES: ***REMOVED***
        prev = buffer.length ? buffer.pop()[1] : ''
        n = css.charCodeAt(pos + 1)
        if (
          prev === 'url' &&
          n !== SINGLE_QUOTE &&
          n !== DOUBLE_QUOTE &&
          n !== SPACE &&
          n !== NEWLINE &&
          n !== TAB &&
          n !== FEED &&
          n !== CR
        ) ***REMOVED***
          next = pos
          do ***REMOVED***
            escaped = false
            next = css.indexOf(')', next + 1)
            if (next === -1) ***REMOVED***
              if (ignore || ignoreUnclosed) ***REMOVED***
                next = pos
                break
          ***REMOVED*** else ***REMOVED***
                unclosed('bracket')
          ***REMOVED***
        ***REMOVED***
            escapePos = next
            while (css.charCodeAt(escapePos - 1) === BACKSLASH) ***REMOVED***
              escapePos -= 1
              escaped = !escaped
        ***REMOVED***
      ***REMOVED*** while (escaped)

          currentToken = ['brackets', css.slice(pos, next + 1), pos, next]

          pos = next
    ***REMOVED*** else ***REMOVED***
          next = css.indexOf(')', pos + 1)
          content = css.slice(pos, next + 1)

          if (next === -1 || RE_BAD_BRACKET.test(content)) ***REMOVED***
            currentToken = ['(', '(', pos]
      ***REMOVED*** else ***REMOVED***
            currentToken = ['brackets', content, pos, next]
            pos = next
      ***REMOVED***
    ***REMOVED***

        break
  ***REMOVED***

      case SINGLE_QUOTE:
      case DOUBLE_QUOTE: ***REMOVED***
        quote = code === SINGLE_QUOTE ? "'" : '"'
        next = pos
        do ***REMOVED***
          escaped = false
          next = css.indexOf(quote, next + 1)
          if (next === -1) ***REMOVED***
            if (ignore || ignoreUnclosed) ***REMOVED***
              next = pos + 1
              break
        ***REMOVED*** else ***REMOVED***
              unclosed('string')
        ***REMOVED***
      ***REMOVED***
          escapePos = next
          while (css.charCodeAt(escapePos - 1) === BACKSLASH) ***REMOVED***
            escapePos -= 1
            escaped = !escaped
      ***REMOVED***
    ***REMOVED*** while (escaped)

        currentToken = ['string', css.slice(pos, next + 1), pos, next]
        pos = next
        break
  ***REMOVED***

      case AT: ***REMOVED***
        RE_AT_END.lastIndex = pos + 1
        RE_AT_END.test(css)
        if (RE_AT_END.lastIndex === 0) ***REMOVED***
          next = css.length - 1
    ***REMOVED*** else ***REMOVED***
          next = RE_AT_END.lastIndex - 2
    ***REMOVED***

        currentToken = ['at-word', css.slice(pos, next + 1), pos, next]

        pos = next
        break
  ***REMOVED***

      case BACKSLASH: ***REMOVED***
        next = pos
        escape = true
        while (css.charCodeAt(next + 1) === BACKSLASH) ***REMOVED***
          next += 1
          escape = !escape
    ***REMOVED***
        code = css.charCodeAt(next + 1)
        if (
          escape &&
          code !== SLASH &&
          code !== SPACE &&
          code !== NEWLINE &&
          code !== TAB &&
          code !== CR &&
          code !== FEED
        ) ***REMOVED***
          next += 1
          if (RE_HEX_ESCAPE.test(css.charAt(next))) ***REMOVED***
            while (RE_HEX_ESCAPE.test(css.charAt(next + 1))) ***REMOVED***
              next += 1
        ***REMOVED***
            if (css.charCodeAt(next + 1) === SPACE) ***REMOVED***
              next += 1
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

        currentToken = ['word', css.slice(pos, next + 1), pos, next]

        pos = next
        break
  ***REMOVED***

      default: ***REMOVED***
        if (code === SLASH && css.charCodeAt(pos + 1) === ASTERISK) ***REMOVED***
          next = css.indexOf('*/', pos + 2) + 1
          if (next === 0) ***REMOVED***
            if (ignore || ignoreUnclosed) ***REMOVED***
              next = css.length
        ***REMOVED*** else ***REMOVED***
              unclosed('comment')
        ***REMOVED***
      ***REMOVED***

          currentToken = ['comment', css.slice(pos, next + 1), pos, next]
          pos = next
    ***REMOVED*** else ***REMOVED***
          RE_WORD_END.lastIndex = pos + 1
          RE_WORD_END.test(css)
          if (RE_WORD_END.lastIndex === 0) ***REMOVED***
            next = css.length - 1
      ***REMOVED*** else ***REMOVED***
            next = RE_WORD_END.lastIndex - 2
      ***REMOVED***

          currentToken = ['word', css.slice(pos, next + 1), pos, next]
          buffer.push(currentToken)
          pos = next
    ***REMOVED***

        break
  ***REMOVED***
***REMOVED***

    pos++
    return currentToken
***REMOVED***

  function back(token) ***REMOVED***
    returned.push(token)
***REMOVED***

  return ***REMOVED***
    back,
    nextToken,
    endOfFile,
    position
***REMOVED***
***REMOVED***
