// detective/index.js
var acorn = require('acorn-node');
var walk = require('acorn-node/walk');
var defined = require('defined');

var requireRe = /\brequire\b/;

function parse (src, opts) ***REMOVED***
    if (!opts) opts = ***REMOVED******REMOVED***;
    var acornOpts = ***REMOVED***
        ranges: defined(opts.ranges, opts.range),
        locations: defined(opts.locations, opts.loc),
        allowReserved: defined(opts.allowReserved, true),
        allowImportExportEverywhere: defined(opts.allowImportExportEverywhere, false)
***REMOVED***;

    // Use acorn-node's defaults for the rest.
    if (opts.ecmaVersion != null) acornOpts.ecmaVersion = opts.ecmaVersion;
    if (opts.sourceType != null) acornOpts.sourceType = opts.sourceType;
    if (opts.allowHashBang != null) acornOpts.allowHashBang = opts.allowHashBang;
    if (opts.allowReturnOutsideFunction != null) acornOpts.allowReturnOutsideFunction = opts.allowReturnOutsideFunction;

    return acorn.parse(src, acornOpts);
***REMOVED***

var exports = module.exports = function (src, opts) ***REMOVED***
    return exports.find(src, opts).strings;
***REMOVED***;

exports.find = function (src, opts) ***REMOVED***
    if (!opts) opts = ***REMOVED******REMOVED***;
    
    var word = opts.word === undefined ? 'require' : opts.word;
    if (typeof src !== 'string') src = String(src);
    
    var isRequire = opts.isRequire || function (node) ***REMOVED***
        return node.callee.type === 'Identifier'
            && node.callee.name === word
        ;
***REMOVED***;
    
    var modules = ***REMOVED*** strings : [], expressions : [] ***REMOVED***;
    if (opts.nodes) modules.nodes = [];
    
    var wordRe = word === 'require' ? requireRe : RegExp('\\b' + word + '\\b');
    if (!wordRe.test(src)) return modules;
    
    var ast = parse(src, opts.parse);
    
    function visit(node, st, c) ***REMOVED***
        var hasRequire = wordRe.test(src.slice(node.start, node.end));
        if (!hasRequire) return;
        walk.base[node.type](node, st, c);
        if (node.type !== 'CallExpression') return;
        if (isRequire(node)) ***REMOVED***
            if (node.arguments.length) ***REMOVED***
                var arg = node.arguments[0];
                if (arg.type === 'Literal') ***REMOVED***
                    modules.strings.push(arg.value);
            ***REMOVED***
                else if (arg.type === 'TemplateLiteral'
                        && arg.quasis.length === 1
                        && arg.expressions.length === 0) ***REMOVED***

                    modules.strings.push(arg.quasis[0].value.raw);
            ***REMOVED***
                else ***REMOVED***
                    modules.expressions.push(src.slice(arg.start, arg.end));
            ***REMOVED***
        ***REMOVED***
            if (opts.nodes) modules.nodes.push(node);
    ***REMOVED***
***REMOVED***
    
    walk.recursive(ast, null, ***REMOVED***
        Statement: visit,
        Expression: visit
***REMOVED***);
    
    return modules;
***REMOVED***;
